[GEP-31] In-Place Node Updates of Shoot Clusters (#10828)
gardener/gardener: docs/proposals/31-inplace-node-update.md

    - [Extensions](#extensions)
      - [OS extensions](#os-extensions)
    - [Failures and Recovery Strategy](#failures-and-recovery-strategy)
      - [Types of Failures During Update](#types-of-failures-during-update)
        - [Drain Timeout/Failure](#drain-timeoutfailure)
        - [Kubernetes Update Failures](#kubernetes-update-failures)
        - [OS Update Failures](#os-update-failures)
      - [Roles in Failure Handling](#roles-in-failure-handling)
    - [Additional Validations](#additional-validations)
  - [Future Work](#future-work)
  - [Alternatives](#alternatives)
    - [Using Only Gardener Node Agent](#using-only-gardener-node-agent)
    - [Using the same name for Machine Class](#using-the-same-name-for-machine-class)

## Summary

Gardener should support new update strategies for nodes that allow updates to Kubernetes minor versions and/or Machine image/OS versions without requiring the deletion and recreation of the nodes. These strategies aim to minimize the overhead traditionally associated with node replacement and offer an alternative approach to updates, particularly important for physical machines or bare-metal nodes.

> The in-place update strategy is not intended to replace the current rolling update strategy provided by the [`machine-controller-manager`](https://github.com/gardener/machine-controller-manager). Instead, it is designed for specific scenarios where physical machines or bare metal nodes—whether manually joined or programmatically provisioned, cannot be easily rolled. Additionally, this update strategy can be useful in scenarios where rolling machines are undesirable, such as when dealing with large machines or limited resources on the cloud provider. For all other scenarios where preserving nodes is not critical, the rolling update strategy remains the preferred approach.

Reviewer: ```suggestion
The in-place update strategy is not intended to replace the current rolling update strategy provided by the [`machine-controller-manager`](https://github.com/gardener/machine-controller-manager). Instead, it is designed for specific scenarios where physical machines or bare metal nodes, whether manually joined or programmatically provisioned, cannot be easily rolled. Additionally, this update strategy can be useful in scenarios where rolling machines are undesirable, such as when dealing with large machines or limited resources on the cloud provider. For all other scenarios where preserving nodes is not critical, the rolling update strategy remains the preferred approach.
```

> The in-place update strategy is not intended to replace the current rolling update strategy provided by the [`machine-controller-manager`](https://github.com/gardener/machine-controller-manager). Instead, it is designed for specific scenarios where physical machines or bare metal nodes, whether manually joined or programmatically provisioned, cannot be easily rolled. Additionally, this update strategy can be useful in scenarios where rolling machines are undesirable, such as when dealing with large machines or limited resources on the cloud provider. For all other scenarios where preserving nodes is not critical, the rolling update strategy remains the preferred approach.

Reviewer 2: As written in the underlying ticket, I would not call one thing "rolling update" and the other thing "in-place update". Both approaches can be (automatically) rolling or not (automatically) rolling (and instead steered by the user which will probably also orchestrate a rolling update in her/his interest). From the ticket:

- TerminateAndRecreateUpdate:
    - Rolling: Changing the spec triggers a rolling update of the nodes according to maxUnavailable and maxSurge (either >0) <-- what we have today
    - OnDelete: Changing the spec doesn’t trigger anything, but new nodes get the new spec; old nodes can be directly deleted
    - OnLabel: Changing the spec doesn’t trigger anything, but new nodes get the new spec; old nodes can be rolled via a label
- InPlaceUpdate:
    - Rolling: Changing the spec triggers an in-place update of the nodes according to maxUnavailable (>0) while maxSurge must be 0 <-- what you describe here
    - OnLabel: Changing the spec doesn’t trigger anything, but nodes that are labelled will be in-place updated  <-- what you describe here

Reviewer 3: How about `RecreateUpdate` instead of `TerminateAndRecreateUpdate`, Recreate itself implies the machine is deleted and created again. So it would be `RecreateUpdateRolling`, `RecreateUpdateOnDelete`, `RecreateUpdateOnLabel`, `InPlaceUpdateRolling`, `InPlaceUpdateOnLabel`.
PS: There is also a strategy called `Recreate` in the MCM, https://github.com/gardener/machine-controller-manager/blob/f17b00e6b63505388f2ceb629b89d43cf5e3603b/pkg/apis/machine/types.go#L489C1-L495C2

Reviewer 2: Sure, sounds good.

On second thought, maybe RecreateUpdateOnDelete is not necessary as additional "strategy".

Reason: The main difference is probably whether it's **user-triggered** (e.g. on label; @unmarshall suggested not to leak the concrete means, in this case, setting a label, into the API and only focus on the fact that the user decides) or **Gardener-triggered** (i.e. automatically rolling), but if people want to "replace" (recreate) a machine by deleting it, they can do that anyway.
So all what the users need to do is tell us whether they want to be in control (user-triggered) or we should be in control (Gardener-triggered). When they are in control, they can decide to use the means we are providing to them (e.g. setting a label) or simply delete a machine explicitly or even let CA do that (e.g. if it doesn't run any/enough pods anymore; maybe aided by taints they applied to the nodes after changing the spec, so that no new pods get scheduled to old machines and they slowly go out of live).

So, it would be rather:
- `RecreateUserTriggered` <-- what HANA and others asked for in the past
- `RecreateGardenerRolling` <-- what we have today
- `InPlaceUserTriggered` <-- what we are planning to do
- `InPlaceGardenerRolling` <-- what we are planning to do

Reviewer 2: After some discussion, we agreed on allowing users to choose between the old replace and the new in-place strategy (in case there are good reasons we do not foresee today - and for backward compatibility resp. giving the users the means to go back to what we had before in case there are issues).

We also agreed on the following names:
- `ManualReplace` <-- what HANA and others asked for in the past
- `AutoReplace` <-- what we have today
- `ManualInPlace` <-- what we are planning to do
- `AutoInPlace` <-- what we are planning to do

It will be allowed to freely transition between `Manual` and `Auto` and vice versa, but...

it will **not** be allowed to transition between `Replace` and `InPlace` or vice versa, because that would complicate the implementation (see hash problem mentioned further down), but it can be added later (`InPlace` to `Replace` is easy, but so is the mitigation for the end user by creating a new worker pool; `Replace` to `InPlace` is more difficult and may require to at least once to recreate the machines and only then further updates can be `InPlace`).

The above will be put into the GEP.

Reviewer 4: A better name for default rolling updates would be `AutoRollingUpdate` instead of using `AutoReplaceUpdate` which is hard to reason. It seems that others like Kibana, ElasticSearch also use the same term.

Reviewer 3: Renamed with https://github.com/gardener/gardener/pull/10828/commits/41cb6e65e01fba522ef683addabee3cabff351ac.

In Gardener, when the Kubernetes minor version and/or Machine image version of a worker pool in a shoot is updated, the nodes of the worker pool are typically rolled. This involves creating new nodes with an updated configuration and deleting the old nodes.

This process presents challenges for **physical machines** or **bare-metal nodes**, which are not as interchangeable as virtual machines because of factors such as:

- **Long boot times**: Physical machines often have longer boot processes, especially when running Power-On Self-Tests (POST), such as memory checks, unless they are fully booted already and can fast-boot.
- **Slow cleanup/sanitization**: Erasing data from physical machines can take significant time before they can be repurposed or returned to a resource pool.
- **Locally attached storage**: Physical machines often have locally attached disks, which cannot be easily moved or re-attached to other machines, unlike network-attached volumes in virtual environments.
- **Limited interchangeability**: Physical machines are typically larger in size, fewer in number, and not as easily replaced or repurposed as virtual machines, making their replacement more challenging.

> 

Reviewer 2: ```suggestion

In addition, this process also presents challenges for **virtual machines**, e.g. if the virtual machine type is scarce (in general or because of an ongoing capacity crunch) or if certain scarce resources are connected/attached, e.g. GPUs that may not be available for additional or even replaced machines once released (as other customers are waiting for them).

> To address this, we aim to support new update strategies that do not require the deletion and recreation of nodes when the Kubernetes minor version and/or machine image version is updated, reducing the need for node replacement—especially in environments with limited physical resources. Additionally, there are scenarios where users want control over node updates, allowing them to select specific nodes to be updated. Our goal is to support a new way to update these dependencies that avoids node deletion and recreation, while also giving users the flexibility to manage updates on particular nodes.

Reviewer: ```suggestion
To address this, we aim to support new update strategies that do not require the deletion and recreation of nodes when the Kubernetes minor version and/or machine image version is updated, reducing the need for node replacement, especially in environments with limited physical resources. Additionally, there are scenarios where users want control over node updates, allowing them to select specific nodes to be updated. Our goal is to support a new way to update these dependencies that avoids node deletion and recreation, while also giving users the flexibility to manage updates on particular nodes.
```

> To address this, we aim to support new update strategies that do not require the deletion and recreation of nodes when the Kubernetes minor version and/or machine image version is updated, reducing the need for node replacement, especially in environments with limited physical resources. Additionally, there are scenarios where users want control over node updates, allowing them to select specific nodes to be updated. Our goal is to support a new way to update these dependencies that avoids node deletion and recreation, while also giving users the flexibility to manage updates on particular nodes.

Reviewer 2: ```suggestion
To address this, we aim to support new update strategies that do not require the deletion and recreation of nodes when the Kubernetes minor version and/or machine image version is updated, reducing the need for node replacement, especially in environments with limited physical resources. Additionally, there are scenarios where users want control over node updates, allowing them to select specific nodes to be updated. Our goal is to support a new way to update these dependencies that avoids node deletion and recreation, while also giving users the flexibility to orchestrate (=trigger) updates themselves as they see fit, e.g. after executing custom pre-update checks.
```

### Goals

> - Provide functionality to update the nodes in-place for Kubernetes minor versions and Machine image versions.

Reviewer 2: ```suggestion
- Provide functionality to update the nodes in-place for Kubernetes minor versions and machine image versions.
```

Reviewer 2: This is not true. Please increase the scope as described by the ticket https://github.com/gardener/gardener/issues/10219. The goal is to cover 100% of the cases where we terminate and recreate machines today, not only for K8s minor versions and OS updates. Whenever we would terminate and recreate (e.g. changing certain today prohibited kubelet settings or during credentials rotation), we have to do it in-place. You can list these two examples, but everything must be in-place - without exception (unless the change is machine-specific, which will never be the case for actual bare metal nodes).

> - Two strategies for in-place - `InPlaceUpdate` and `InplaceUpdateOnLabel`.

Reviewer 2: See above. Instead of differentiating both by omitting "OnLabel", maybe call them `InPlaceUpdateRolling` and `InPlaceUpdateOnLabel`.

The word rolling doesn't imply adding new nodes before removing them. See e.g. the current MCM update: it allows `maxUnavailable` > 0 and `maxSurge` = 0, which is not adding new nodes before taking down existing ones.

Rather the differentiation is whether some automaton will go through the nodes (rolling) or the user will define which ones will be updated and when (on label).

Also, please notice the upper case `P` in `InPlace`. This document inconsistently writes `InPlaceUpdate` with an upper case `P` and `InplaceUpdateOnLabel` with a lower case `p`. I will not mention it again, but it should be changed everywhere (I prefer upper case `P`, since it's a compound word).

> - Maximize re-use of the existing functionality from `machine-controller-manager` and [`gardener-node-agent`](https://github.com/gardener/gardener/blob/master/docs/concepts/node-agent.md).

Reviewer 2: This is no goal. This may be an design/implementation decision, but no goal per se.

### Non-Goals

- Provide a way to schedule the pods on the same node after the in-place update.
- Downgrading of Kubernetes/Machine image version after successful update.

> - Switching between in-place and rolling update strategies.

Reviewer 2: Why not? Why would this be an immutable setting? How does that help or simplify things if we force people to recreate new pools?

Author: For the first phase of the implementation i.e. this GEP, we don't want to support switching between strategies as it has a lot of complexity(for ex - OSC hash calculation and GNA secret hash calculation).

Reviewer 2: OK, thank you. On second thought, instead of making it configurable we may also decide later to always do in-place updates. I mean, people cannot opt out of doing K8s patch updates in place, can they? What we can do in-place, we already do in-place and people love it. With in-place updates they will have even less worries, so why would people decide for the recreation of machines if we can update them in-place?

Maybe that's nothing we want to offer later, because nobody would use it anyway? Or is maybe updating the OS something that there would be good reasons for not to do in-place, e.g. when they have a daemonset that is "massaging" the machine somehow? But even then, they have to make their daemonset react accordingly (also today they must deal with a reboot).

What people certainly want to influence is whether it's user-triggered (e.g. on label) or Gardener-triggered (i.e. automatically rolling). And that should be configurable / not conflict with e.g. hash calculation, right?

Reviewer 3: > What people certainly want to influence is whether it's user-triggered (e.g. on label) or Gardener-triggered (i.e. automatically rolling). And that should be configurable / not conflict with e.g. hash calculation, right?

Yes, this is supported, will explicitly mention it.

- Provide an update strategy without draining the node.
- Provide in-place updates only for the Kubernetes version in worker pools where the OS does not support in-place updates.

## Proposal

In short, Gardener should support new update strategies that do not require deleting and recreating the node when the Kubernetes minor or Machine image version is updated.

### Approach

The `machine-controller-manager` and `gardener-node-agent` components will be used to perform in-place updates. `machine-controller-manager` will provide an interface that prepares the machine/node for an update. This interface can be used by other components to determine if a node is ready for an update, allowing the other component to perform the update on the node and then notify the `machine-controller-manager` that the node has been successfully updated. _In the case of Gardener, this component will be the `gardener-node-agent`._

> This design was chosen to leverage the `machine-controller-manager`'s node rollout orchestration functionality while ensuring that it remains a standalone component.

Reviewer 2: Here it's fine. No goal, but a design/implementation decision.

### Prerequisites

The only prerequisite is that the machine image/OS on the node must support in-place updates and provide a tool or utility to initiate the OS update. This update tool/utility should meet the following requirements:

1. **Failure Recovery**: In case of a failure, the OS should be able to revert to the previous version.
2. **Retry Configuration**: Support for configuring the number of update retries.
3. **Registry Configuration**: Ability to specify the registry from which to pull the OS image.

### Update Strategies

> Gardener will introduce two additional update strategies, `InPlaceUpdate` and `InPlaceUpdateOnLabel`, for worker pools, in addition to the existing `RollingUpdate`. The two new strategies are described as follows:

Reviewer 2: As said in the call we had out-of-band and also above, I strongly suggest to improve/change the naming before you start with the implementation. Today, we only call what we have "rolling update" because we only implemented a single procedure and therefore it wasn't necessary to describe it with more words, but today's "rolling update" has neither to remain "rolling" nor does in-place exclude being "rolling".

What we have is rather:
- `TerminateAndRecreateUpdateRolling`

With the option to later add (as part of another GEP):
- `TerminateAndRecreateUpdateOnDelete`
- `TerminateAndRecreateUpdateOnLabel`

What you want to build in the context of this GEP:
- `InPlaceUpdateRolling`
- `InPlaceUpdateOnLabel`

Let's improve the naming now while it's still cheap (editing text is cheap) and before we start with the implementation (editing code is expensive).

Reviewer 5: @vlerenc  apologies, but don't see why we need to rename our current strategy. As of today, our `MachineDeploymentStrategyType` of `RollingUpdate` satisfies the k8s rolling update of no downtime (if configured correctly). We are using the same semantics are the k8s `Deployment` whose `Spec.Strategy.Type` is also worded as `RollingUpdate`.  And we have mostly followed the same semantics.  People who see the same word will expect the same semantics.

After all, for a k8S `Deployment`, `Pods` are also terminated and recreated, but they still call it `RollingUpdate` and not `TerminateAndRecreateUpdateRolling` . It makes sense to stick to the same literal word.  

No objections to the revised words for the strategies: `InPlaceUpdateRolling` and `InPlaceUpdateOnLabel`.  Though to be honest, applying the same logic, even the latter can include the word `Rolling`, right ?   Following same logic, it should be `InPlaceRollingUpdate` and `InPlaceRollingUpdateOnLabel`.   At-least it flows off the tongue better that way in English and is more readable.

Reviewer 2: So much to say:
1.) I like to avoid near-useless renaming effort as well.
2.) Then again, good names, especially if exposed to end-users in the Gardener API, make our lives easier later on. So far, the `MachineDeploymentStrategyType` is only internal (MCM API) - we can keep it/map it. But exposing now strategies with "too wide/ambiguous" names to end users in our Gardener API seems like something we should like to disambiguate/make more clear if possible.
3.) "Deployments" and "MachineDeployments" are not the same thing, e.g. because pods aren't updated in place (even changing the requests in-place was a difficult thing), but machines soon will be (which incidentally is why we are in this situation now with "MachineDeployments" and Kubernetes isn't with "Deployments").
4.) But the main point really is that "MachineDeployment"s and the whole thought of "similarity" to which you spoke is hidden at the Gardener API. Hence, no user will make the connection anyway. As far as they see it, they are asked to define the strategy for how to update workers in a worker pool.
5.) As for "rolling" in combination with "recreate" or "inplace", you cannot know what the user is up to. Maybe she/he just wants to take us out of the picture and there is no real rolling at all - it's more a blue-green update they steer with the deletion of a machine or its label (new nodes are added, but old nodes were tainted and will go out of live over time). We should not assume, users will do the same as we do when we roll (recreate or in-place), when they are in control. We can only roll, because that's all we know how to do, but users may have other ideas what to do with the means we expose to them. 

So, when this GEP writes "Gardener will introduce [] additional update strategies..." I see the Gardener API (this is the gardener repo after all), but you possibly the MCM API? At the Gardener API level I really like to avoid now introducing a strategy name (newly), just because we introduced it in former years at the MCM level (probably needlessly back then, because there was no other strategy anyhow) that nobody ever saw at Gardener API level.

So, at Gardener API level, all strategies are new, so let's give them good names, please. That's no renaming for me - it's exposing a completely new API for worker update strategies in the shoot API.

#### InplaceUpdate strategy

> 1. On shoot spec change for Kubernetes/Machine image version, `machine-controller-manager` will label all the nodes with `candidate-for-update` label. This label can be used to identify nodes that will be selected for updates in the future.

Reviewer 2: ```suggestion
1. On a shoot spec change for a worker pool, `machine-controller-manager` will label all nodes of the changed worker pool with the `candidate-for-update` label. This label is used to identify nodes that require an update.
```

Reviewer 2: Will this suffice? Or will it become necessary to use labels with values expressing a checksum (e.g. of the worker pool configuration) or similar, so that it's clear what needs to be updated, e.g. if the user makes new changes while the last changes haven't been rolled out yet (or are still in progress)?

At a minimum there is the case that while the node agent updates a node, the user changes the spec again. How is this handled (safely)?

> 2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent it from being scaled down by [`cluster-autoscaler`](https://github.com/gardener/autoscaler/tree/machine-controller-manager-provider/cluster-autoscaler) during the update process.

Reviewer 2: ```suggestion
2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent them from being scaled down by [`cluster-autoscaler`](https://github.com/gardener/autoscaler/tree/machine-controller-manager-provider/cluster-autoscaler) during the update process.
```

> 3. `machine-controller-manager` will select the nodes based on the configured `maxUnavailable` and label them with `selected-for-update`. This label can be used to identify nodes that are currently undergoing an update.

Reviewer 2: ```suggestion
3. `machine-controller-manager` will select as many nodes as are permitted by the `maxUnavailable` setting for the worker pool and label it/them with `selected-for-update`. This label can be used to identify nodes that are about to or are currently undergoing an update.
```

> 4. `machine-controller-manager` will cordon and drain the node and label it with `ready-for-update` label.

Reviewer 2: ```suggestion
4. `machine-controller-manager` will cordon and drain the node/nodes and label it/them with the `ready-for-update` label once the drain has completed.
```

> 5. `gardener-node-agent` will detect the `ready-for-update` label and perform the update on the node. And it will wait for the update to complete within the specified timeout.

Reviewer 2: ```suggestion
5. `gardener-node-agent` will detect the `ready-for-update` label and perform the update on the machine. It will wait for the update to complete (with a specified timeout).
```

> 6. Once the Kubernetes/Machine image version is updated, the node will be labeled by the `gardener-node-agent` with `update-successful` label.

Reviewer 2: ```suggestion
6. Once the machine has been updated, the node will be labeled by the `gardener-node-agent` with the `update-successful` label.
```

> 7. `machine-controller-manager` will `uncordon` the node, and remove all the added labels. The node is then ready to schedule the workloads.

Reviewer 2: ```suggestion
7. `machine-controller-manager` will `uncordon` any node with the `update-successful` label and remove this and all other update-related labels. When the node becomes `Ready`, it can host again workload.
```

![InplaceUpdate](assets/inplace-update.png "InPlace Update Strategy")

#### InplaceUpdateOnLabel strategy

> 1. On shoot spec change, nodes will be labeled with `candidate-for-update`.

Reviewer 2: See above:

```suggestion
1. On a shoot spec change for a worker pool, `machine-controller-manager` will label all nodes of the changed worker pool with the `candidate-for-update` label. This label is used to identify nodes that require an update.
```

> 2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent it from being scaled down by `cluster-autoscaler` during the update process.

Reviewer 2: See above:

```suggestion
2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent them from being scaled down by [`cluster-autoscaler`](https://github.com/gardener/autoscaler/tree/machine-controller-manager-provider/cluster-autoscaler) during the update process.
```

Reviewer 3: The URL is already added for `cluster-autoscaler` in the above section, should we add them again?

Reviewer 2: No need to have the URL a second time (not the essence of this change request), but it also doesn't hurt (I am totally neutral).

My main reason for the change was the grammar, i.e. changing "it" -> "them" because of the plural.

The URL came (side-effect) because I copied the section from above that was grammatically correct. That's all.

> 3. `machine-controller-manager` will wait for the `selected-for-update` label on the node. The orchestration of the update is solely the responsibility of the user. They are free to select the nodes to be updated at will.

Reviewer 2: ```suggestion
3. `machine-controller-manager` will wait for the `selected-for-update` label on the node. The orchestration of the update is solely the responsibility of the user. The user is free to select the nodes to be updated at will.
```

> 4. Once the `selected-for-update` label is added by the user on the node, steps 4-7 of the `Inplace` strategy will performed.

Reviewer 2: ```suggestion
4. Once the `selected-for-update` label is added by the user, steps 4-7 of the `InplaceUpdateRolling` strategy will be performed.
```

![InplaceUpdateOnLabel](assets/inplace-update-onlabel.png "InPlace Update OnLabel Strategy")

### Gardener

#### `CloudProfile` API

> A new field `minSupportedVersionInPlaceUpdate` is introduced under `.spec.machineImages[]` under the CloudProfile spec. This field can be used to specify the minimum supported version from which the current machine image version can be in-place updated to.

Reviewer 2: ```suggestion
A new field `minSupportedVersionInPlaceUpdate` is introduced under `.spec.machineImages[].version[]` in the `CloudProfile`. This field will be used to express possible upgrade constraints, specifically the minimum version required to upgrade to the new version defined in the `CloudProfile`. 

Here's how it works:
1. The `minSupportedVersionInPlaceUpdate` field is used to set the lowest version that can be directly upgraded to the new version.
2. If a worker pool is running a version that is even lower than the `minSupportedVersionInPlaceUpdate`, it won't be able to directly upgrade to the new version.
3. Instead, the system will calculate and orchestrate a series of intermediate upgrades. This means the worker pool will first be upgraded to an intermediate version (or versions) before it can be upgraded to the new version.
```

...maybe you had something else in mind than (3), then write that, e.g. prevent updating to a new version if the constraint would be violated (admission plugin).

Author: A new field `minSupportedVersionInPlaceUpdate` has been introduced under `.spec.machineImages[].versions[]` in the `CloudProfile`. This field specifies the lowest machine image version that can be directly upgraded to the new version.

How about changing it to the above line? Point 1-2 explain the same thing in two ways, also how this will be done(ex through the admission plugin) is more of implementation details and should not be mentioned in the GEP IMO.

Reviewer 5: Minor nit: not sure why `supported` adjective is necessary in the above. It is implied implicitly.  Why can't it simply be `minVersionForInPlaceUpdate`

Reviewer 2: In any case, please explain whether "we" ("the system") will calculate "intermediate upgrades" or not and the user has to find out and "plan" the "intermediate upgrades".

That was the main point of this review comment, because it wasn't clear to me when reading the text. I therefore wrote:
> ...maybe you had something else in mind than (3), then write that, e.g. prevent updating to a new version if the constraint would be violated (admission plugin).

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: CloudProfile
metadata:
  name: cloudprofile1
spec:
  type: <some-provider-name>
  providerConfig: <some-provider-specific-cloudprofile-config>
  kubernetes:
    versions:
      - version: 1.31.1
  machineImages:
    - name: gardenlinux
      versions:
        - version: 1443.8.0
          expirationDate: "2025-02-28T23:59:59Z"
          minSupportedVersionInPlaceUpdate: 1312.3.0
        - version: 1443.7.0
  machineTypes:
    - name: m5.large
      cpu: "2"
      gpu: "0"
      memory: 8Gi
      usable: true
  regions:
    - name: <region-name>
```

#### `Shoot` API

> A new field `UpdateStrategy` is introduced under `spec.provider.workers[]` in the Shoot spec. This field will be passed on to the worker extension. Once set, changing the `UpdateStrategy` is prohibited through validation. Additionally, when the `InPlaceUpdate/InplaceUpdateOnLabel` update strategy is configured, skipping the worker pool's Kubernetes minor version is not allowed.

Reviewer 6: Is there a specific reason why minor version skipping will not be supported with in-place updates?

Reviewer 3: Skipping minor versions is generally not recommended since there could be breaking changes or required conversions between versions. However, in rolling updates, the node is replaced anyway, so we allowed skipping minor versions with https://github.com/gardener/gardener/issues/9184.

Reviewer 2: > Once set, changing the `UpdateStrategy` is prohibited through validation.

As written above, I lack the reasoning for this restriction (was a non-goal). Why is this necessary? How does it make anything simpler?

Reviewer 2: In addition, since @dguendisch asked, others may ask the same question, so let's give them your answer right away.

> A new field `UpdateStrategy` is introduced under `spec.provider.workers[]` in the Shoot spec. This field will be passed on to the worker extension. Once set, changing the `UpdateStrategy` is prohibited through validation. Additionally, when the `InPlaceUpdate/InplaceUpdateOnLabel` update strategy is configured, skipping the worker pool's Kubernetes minor version is not allowed.

Reviewer 2: ```suggestion
A new field `updateStrategy` is introduced under `spec.provider.workers[]` in the Shoot spec. This field will be passed on to the worker extension. Once set, changing the `updateStrategy` is prohibited through validation. Additionally, when `InPlaceUpdateRolling` or `InPlaceUpdateOnLabel` is configured as update strategy, skipping an intermediate Kubernetes minor version is no longer allowed, because of potential local changes that are then not carried out anymore when skipping versions (only allowed with `TerminateAndRecreateUpdateRolling` as the machines/nodes are created from scratch in this case).
```

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
metadata:
  name: crazy-botany
  namespace: garden-dev
spec:
  secretBindingName: my-provider-account
  cloudProfile:
    name: cloudprofile1
  region: europe-central-1
  provider:
    type: <some-provider-name> # {aws,azure,gcp,...}
    workers:
      - name: cpu-worker
        minimum: 5
        maximum: 5
        maxSurge: 0
        maxUnavailable: 2

>         updateStrategy: InPlaceUpdate # RollingUpdate/InPlaceUpdate/InplaceUpdateOnLabel

Reviewer 2: ```suggestion
        updateStrategy: InPlaceUpdateRolling # TerminateAndRecreateUpdateRolling, InPlaceUpdateRolling, or InPlaceUpdateOnLabel
```

```yaml
apiVersion: extensions.gardener.cloud/v1alpha1
kind: Worker
metadata:
  name: worker
  namespace: shoot--foobar--aws
spec:
  type: aws
  region: eu-west-1
  secretRef:
    name: cloudprovider
    namespace: shoot--foobar--aws
  infrastructureProviderStatus: <infra-status>
  pools:
    - name: cpu-worker-1
      machineType: m4.large
      machineImage:
        name: gardenlinux
        version: 1.2.2
      minimum: 5
      maximum: 5
      maxSurge: 0
      maxUnavailable: 2
      updateStrategy: InPlaceUpdate # RollingUpdate/InPlaceUpdate/InplaceUpdateOnLabel
```

#### `OperatingSystemConfig` API

A new field `version` is introduced in the `OperatingSystemConfig` spec. It specifies the machine image/OS version of the worker pool.

```yaml
apiVersion: extensions.gardener.cloud/v1alpha1
kind: OperatingSystemConfig
metadata:
  name: pool-01-original
  namespace: default
spec:
  type: <os-type>
  version: <operating-system-version>
  units:
    - ...
  files:
    - ...
```

#### Gardener Node Agent

> The `gardener-node-agent` will function as the tool responsible for updating Kubernetes minor version and Machine image version dependencies on the node.

Reviewer 2: Again, we ought to carry out 100% of all possible update cases that do not involve the physical/virtual machine in-place. The GEP is triggered by the/this requirement to carry out any changes in-place, not just Kubernetes minor version and machine image version upgrades.

Also (commented only now, but corrected multiple times above), why is machine often written with an upper case `M`? Kubernetes is a name (so upper case 'K'), but machine isn't (so lower case 'm') - this is not the `Machine` resource in the seed cluster/as managed by MCM, but reads like the general word "machine" (not `Machine` in code quotes).

> Currently, when a dependency changes in the Operating System Configuration (OSC), `gardener-node-agent` detects the difference between the current OSC and the last applied OSC, and updates any changed unit on the node. When the in-place update strategy is configured, and either the Kubernetes minor version or the machine image version is changed, `gardener-node-agent` will not immediately apply the updates to the node. Instead, it will wait for the `ready-to-update` label on the node. Once the label is present, `gardener-node-agent` will proceed with the update.

Reviewer 2: Again, this GEP cannot be only about Kubernetes or machine image versions:

```suggestion
Currently, when a dependency changes in the Operating System Configuration (OSC), `gardener-node-agent` detects the difference between the current OSC and the last applied OSC, and updates any changed unit on the node. When the in-place update strategy is configured, and anything changes that may not be changed in-place right away (such as the Kubernetes minor version, machine image version, credentials rotation i.e. a new CA, etc.), `gardener-node-agent` will not immediately apply the updates to the node. Instead, it will wait for the `ready-to-update` label on the node. Once the label is present, `gardener-node-agent` will proceed with the update.
```

In the case of OS updates, `gardener-node-agent` expects an "update-script" to be present in the `.spec.files[]` of the OSC at a predefined path, and the same will be executed to perform the update.

After the update, based on its success or failure, `gardener-node-agent` will add the appropriate label(`update-successful` or `update-failed`) to the node to indicate the status of the update.

### Machine Controller Manager

The `machine-controller-manager` acts as an orchestrator for the update, ensuring changes are rolled out in a controlled manner.

> In the current implementation, when a Kubernetes minor version or Machine Image/OS version is updated in the shoot (worker pool), the provider extension creates a new `MachineClass`, and the `MachineDeployment` is updated to reference this new `MachineClass`. `machine-controller-manager` detects this change and creates a new `MachineSet` with the new `MachineClass` and machines are rolled out based on the `RollingUpdate` strategy.

Reviewer 2: If I am not mistaken, that's not hardcoded, but checksum-based? Certain changes affect the resources today, certain others don't. In the future, none may require the creation of new machines unless the actual physical/virtual machine is changed, I think.

Reviewer 3: > If I am not mistaken, that's not hardcoded, but checksum-based?

Yes. Any change which needs the recreation of a machine will need to generate a new checksum.

Reviewer 2: Yes, my point is merely that probably in the future, most changes will be automatically carried out in-place, because why not? Whatever we can do in-place we should do in-place and if only the machine type remains as hard reason to throw away machines, then this will be great news for our end users, I think.

> In the case of the `RollingUpdate` strategy, new machines are created in the new `MachineSet` using the new `MachineClass`, and the old ones are deleted from the old `MachineSet`. For in-place updates, instead of creating new machines for the new `MachineSet`, the old ones are moved from the old `MachineSet` to the new one after the `gardener-node-agent` has successfully updated the dependencies. This process preserves the old nodes.

Reviewer 2: ```suggestion
In the case of the `RollingUpdate` strategy, new machines are created in the new `MachineSet` using the new `MachineClass`, and the old ones are deleted from the old `MachineSet`. For in-place updates, instead of creating new machines for the new `MachineSet`, the old ones are moved from the old `MachineSet` to the new one after the `gardener-node-agent` has successfully updated the dependencies. This process preserves the old machines (their `Machine` and `Node` resources).
```

There are two ways to achieve this:

> 1. **Owner Reference Update**: Move the old machine to the new `MachineSet` by updating the owner reference. The downside of this method is that the machine name will stay the same through updates, which breaks the current naming convention of including the `MachineSet` name as a prefix in machine names.

Reviewer 2: Here it's the `Machine` (in code quotes):

```suggestion
1. **Owner Reference Update**: Move the old `Machine` to the new `MachineSet` by updating the owner reference. The downside of this method is that the `Machine` name will stay the same through updates, which breaks the current naming convention of including the `MachineSet` name as a prefix in `Machine` name.
```

> 2. **Shallow Delete**: Instead of directly moving a machine from one `MachineSet` to another, first perform a shallow delete of the old machine object (without deleting the underlying VM), then create a new machine object in the new `MachineSet` for the same VM.

Reviewer 2: ```suggestion
2. **Shallow Delete**: Instead of directly moving a `Machine` from one `MachineSet` to another, first perform a shallow delete of the old `Machine` object (without deleting the underlying physical/virtual machine), then create a new `Machine` object in the new `MachineSet` (for the same physical/virtual machine).
```

> [!NOTE]
> At the time of writing this GEP, it is undecided which method to proceed with. The first method has been evaluated, but the feasibility of the second method is yet to be assessed. If feasible, the second method is preferred.

#### `MachineDeployment` API

A new strategy type `InPlaceUpdate` and a new field, `inPlaceUpdate`, have been introduced under the `.spec.strategy` field. `inPlaceUpdate` can be configured when the selected strategy type is `InPlaceUpdate`.

```yaml
apiVersion: machine.sapcloud.io/v1alpha1
kind: MachineDeployment
metadata:
  name: <name>
  namespace: <shoot-namespace>
spec:
  replicas: 1
  selector:
    matchLabels:
      name: <label>
  strategy:
    inPlaceUpdate:
      maxUnavailable: 1
      maxSurge: 0
      onLabel: false # true/false
    type: InPlaceUpdate # RollingUpdate/InPlaceUpdate/Recreate
  template: {}
    spec: {}
status: {}

> ```

Reviewer 2: Please adapt to the proposals above in regards to the naming, current and future update strategies.

`maxUnavailable`: Specifies the maximum number of unavailable machines during the update process.

> `maxSurge`: Specifies the maximum number of machines that can be scheduled above the desired number of machines. Set `maxSurge` to zero if you don't want any new nodes to be provisioned during the update. You can set it to a value greater than zero if you don't want unavailable nodes during the update; in that case, first new nodes will be created with the updated configuration and then the old nodes will undergo in-place updates.

Reviewer 2: ```suggestion
`maxSurge`: Specifies the maximum number of machines that can be scheduled above the desired number of machines. Set `maxSurge` to zero if you don't want any new machines/nodes to be provisioned during the update. You can set it to a value greater than zero if you don't want unavailable nodes during the update; in that case, first new machines/nodes will be created with the updated configuration and then the old machines/nodes will undergo in-place updates.
```

`onLabel`: If the update strategy is `InPlaceUpdateOnLabel` in the worker spec, the worker controller will set this field to true.

### Dependency Watchdog

> Currently, the [`dependency-watchdog`](https://github.com/gardener/dependency-watchdog) monitors the leases of the nodes in the shoot cluster. If nodes more than a configured threshold are unhealthy, it scales down the `machine-controller-manager`, `kube-controller-manager`, and `cluster-autoscaler` to prevent a meltdown. During an ongoing in-place update, DWD can make use of the labels added by `machine-controller-manager` during the `in-place` update to exclude nodes undergoing updates from its health check.

Reviewer 2: ```suggestion
Currently, the [`dependency-watchdog`](https://github.com/gardener/dependency-watchdog) monitors the leases of the nodes in the shoot cluster. If more nodes than a configured threshold are unhealthy, it scales down the `machine-controller-manager`, `kube-controller-manager`, and `cluster-autoscaler` to prevent a meltdown. During an ongoing in-place update, DWD can make use of the labels added by `machine-controller-manager` during the `in-place` update to exclude nodes undergoing updates from its health check.
```

### Failures and Recovery Strategy

In-place updates may fail because of several factors ranging from network issues to compatibility problems. This section outlines the failure scenarios, error classifications, and the specific roles `gardener-node-agent` and `machine-controller-manager` play in failure handling and recovery to ensure minimal disruption.

#### Types of Failures During Update

##### Drain Timeout/Failure

- A node is marked as `ready-for-update` only after a successful drain.
- We use the existing [drain logic](https://github.com/gardener/machine-controller-manager/tree/v0.54.0/pkg/util/provider/drain) provided by the `machine-controller-manager`.
- If a drain operation times out or fails, the node will be forcefully drained.

##### Kubernetes Update Failures

- **Rollback on Failure:** If a Kubernetes version update fails, `gardener-node-agent` will roll back the `kubelet` to the version specified in the last applied `OperatingSystemConfig`. This ensures that the node remains operational.

>   - **Update Halt on Failure:** `gardener-node-agent` will label the node with an `update-failed` label. This label signals `machine-controller-manager` to stop further updates.

Reviewer 2: So any one/single issue blocks the progress? Maybe it should be rather `maxUnavailable`? Not sure... 🤔

Reviewer 4: typically you define an `errorThreshold` during an update. If that gets breached then you stop further updates. `maxUnavailable` can also be used but it will then get overloaded. Its purpose will be two-fold then - one to determine how many number of nodes can be in-place updated concurrently and also to determine when to stop further updates. If you wish to have different threshold value than number of concurrent updates then you cannot just use `maxUnavailable`.

Reviewer 5: Request that `update-failed` have a succinct, error message for the label value that is logable by the MCM and be capable of being put in the `Machine` status.  If this is not ideal, please consider an the use of a separate annotation `node.machine.sapcloud.io/update-message` that can hold the full error message.

Author: > Maybe it should be rather maxUnavailable.

It will be maxUnavailble only, MCM will pick a machine to get an update only if it can fulfill `maxUnavailable`, also any machine undergoing an update or already failed an update will be counted in the unhealthy machines.

Reviewer 2: Then will you/have you updated the sentence, so that we can close this discussion?

Reviewer 3: Yes, we have updated it to https://github.com/gardener/gardener/blob/4242ed67a2914d848346044d66ae1a66c403bdae/docs/proposals/31-inplace-node-update.md?plain=1#L381

##### OS Update Failures

- **Before rebooting for update** - Update cannot start if the machine image version is unavailable, network issues prevent the image pull, or disk storage is insufficient (although proper garbage collection should prevent this); classified as retriable or non-retriable.
- **Failed to boot the new OS** - If the node fails to boot into the new OS version and reverts to the previous version, `gardener-node-agent` will annotate it to mark an update attempt, but the root cause may remain unlogged as filesystems mount later in the booting process; classified as non-retriable.

> - **Total failure** - In cases where the system fails to boot into either of the OS versions, the update attempt will timeout and be marked as failed; classified as non-retriable.

Reviewer 6: Who will mark the node as failed in this case (failure to boot either OS version)?
The diagram suggests the GNA to label the node with `update-failed`, however if the OS cannot boot, GNA will certainly not run.

Author: It would be MCM adding this label, this is mentioned under MCM section - https://github.com/gardener/gardener/blob/1e5bcac44c77df3efafdf59aa93aa71e64debe8e/docs/proposals/31-inplace-node-update.md?plain=1#L351

  To resolve the issue, the error must be fixed manually, and the `update-failed` label should be removed from the node. This allows `machine-controller-manager` to proceed with the update process.

> [!NOTE]
> All mentioned timeouts and retries are subject to change, with exact values to be decided during implementation.

### Additional Validations

In addition to Kubernetes or Machine image updates, several other factors can trigger the rolling updates of nodes in the shoot clusters. These include:

1. Modification of the following fields in the Shoot specification:
   - `.spec.provider.workers[].machine.type`
   - `.spec.provider.workers[].volume`
   - `.spec.provider.workers[].cri`
   - `.spec.systemComponents.nodeLocalDNS`
2. Rotation of the `CA` or `ServiceAccount` signing key.
3. Changes in the `kubelet` configuration (for eg., kubeReserved).

> In-place updates will support (2) and (3) but will disallow (1) with validations in the Shoot API. This ensures that certain critical changes, such as machine type or volume adjustments, follow the traditional rolling update process in which nodes are replaced.

Reviewer 2: OK, now you write that (2) and (3) will be handled in-place, but throughout the entire document you focussed solely on Kubernetes minor and machine image versions. Please rectify that/avoid narrowing down in-place updates only to Kubernetes minor and machine image versions for the first 90% of this GEP only to reveal that all cases will be supported at the end of the document. Mention that practically everything must be handled in-place that isn't a change related to the physical/virtual machine itself (as described by (1)).

    - [Gardener](#gardener)
      - [`CloudProfile` API](#cloudprofile-api)
      - [`Shoot` API](#shoot-api)
      - [`Worker` API](#worker-api)
      - [`OperatingSystemConfig` API](#operatingsystemconfig-api)
      - [Gardener Node Agent](#gardener-node-agent)
    - [Machine Controller Manager](#machine-controller-manager)
      - [`MachineDeployment` API](#machinedeployment-api)
    - [Dependency Watchdog](#dependency-watchdog)
    - [Extensions](#extensions)
      - [OS extensions](#os-extensions)
    - [Failures and Recovery Strategy](#failures-and-recovery-strategy)
      - [Types of Failures During Update](#types-of-failures-during-update)
        - [Drain Timeout/Failure](#drain-timeoutfailure)
        - [Kubernetes Update Failures](#kubernetes-update-failures)
        - [OS Update Failures](#os-update-failures)
      - [Roles in Failure Handling](#roles-in-failure-handling)
    - [Additional Validations](#additional-validations)
  - [Future Work](#future-work)
  - [Alternatives](#alternatives)
    - [Using Only Gardener Node Agent](#using-only-gardener-node-agent)
    - [Using the same name for Machine Class](#using-the-same-name-for-machine-class)

## Summary

> Gardener should support new update strategies for nodes that allow updates to Kubernetes minor versions and/or Machine image/OS versions without requiring the deletion and recreation of the nodes. These strategies aim to minimize the overhead traditionally associated with node replacement and offer an alternative approach to updates, particularly important for physical machines or bare-metal nodes.

Reviewer 7: Do you want to link to the full list of triggers for rolling the nodes (https://github.com/gardener/gardener/blob/master/docs/usage/shoot-operations/shoot_updates.md#rolling-update-triggers)?

### Goals

- Provide functionality to update the nodes in-place for Kubernetes minor versions and machine image versions.
- Two strategies for in-place - `InPlaceUpdate` and `InplaceUpdateOnLabel`.
- Maximize re-use of the existing functionality from `machine-controller-manager` and [`gardener-node-agent`](https://github.com/gardener/gardener/blob/master/docs/concepts/node-agent.md).

### Non-Goals

- Provide a way to schedule the pods on the same node after the in-place update.
- Downgrading of Kubernetes/Machine image version after successful update.
- Switching between in-place and rolling update strategies.
- Provide an update strategy without draining the node.

> - Provide in-place updates only for the Kubernetes version in worker pools where the OS does not support in-place updates.

Reviewer 7: Should we add that we do not intend (initially) to address other triggers for node rolling (see https://github.com/gardener/gardener/blob/master/docs/usage/shoot-operations/shoot_updates.md#rolling-update-triggers) except for Kubernetes minor and machine image version updates?

### Approach

The `machine-controller-manager` and `gardener-node-agent` components will be used to perform in-place updates. `machine-controller-manager` will provide an interface that prepares the machine/node for an update. This interface can be used by other components to determine if a node is ready for an update, allowing the other component to perform the update on the node and then notify the `machine-controller-manager` that the node has been successfully updated. _In the case of Gardener, this component will be the `gardener-node-agent`._

This design was chosen to leverage the `machine-controller-manager`'s node rollout orchestration functionality while ensuring that it remains a standalone component.

### Prerequisites

The only prerequisite is that the machine image/OS on the node must support in-place updates and provide a tool or utility to initiate the OS update. This update tool/utility should meet the following requirements:

1. **Failure Recovery**: In case of a failure, the OS should be able to revert to the previous version.
2. **Retry Configuration**: Support for configuring the number of update retries.

> 3. **Registry Configuration**: Ability to specify the registry from which to pull the OS image.

Reviewer 7: This all sounds very generic so that any operating system could implement it. Do we actually intend to have a generic interface? If yes, how would it look like. If no, let's state that clearly here as well.

For example, if we first only want to make it work with GardenLinux we may want to state that explicitly. Later on, we can still generalise it. However, I would like to understand what we want to do as part of this GEP and what might be future work, which could also be mentioned in the future work section.

Reviewer 3: > This all sounds very generic so that any operating system could implement it.

If they have a mechanism to boot into the new version by triggering some tool/utility, then yes, they can.
It is the choice of the OS, the GEP itself doesn't differentiate between any OS.

> Do we actually intend to have a generic interface?

Isn't the current approach generic enough?

Reviewer 7: With the addition of the update command, it is now clear that any operating system extension could implement the corresponding interface.
Thanks for clarifying.

### Update Strategies

Gardener will introduce two additional update strategies, `InPlaceUpdate` and `InPlaceUpdateOnLabel`, for worker pools, in addition to the existing `RollingUpdate`. The two new strategies are described as follows:

#### InplaceUpdate strategy

1. On a shoot spec change for a worker pool, `machine-controller-manager` will label all nodes of the changed worker pool with the `candidate-for-update` label. This label is used to identify nodes that require an update.
2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent them from being scaled down by [`cluster-autoscaler`](https://github.com/gardener/autoscaler/tree/machine-controller-manager-provider/cluster-autoscaler) during the update process.
3. `machine-controller-manager` will select as many nodes as are permitted by the `maxUnavailable` setting for the worker pool and label it/them with `selected-for-update`. This label can be used to identify nodes that are about to or are currently undergoing an update.

> 4. `machine-controller-manager` will cordon and drain the node/nodes and label it/them with the `ready-for-update` label once the drain has completed.

Reviewer 8: What about pods that are not evicted by a typical drain operation (daemon sets, those with local storage)? Don't they have to be terminated as well for the update? How will this be done?

Author: GNA will restart all the remaining pods on the node after a successful update.

> 5. `gardener-node-agent` will detect the `ready-for-update` label and perform the update on the machine. It will wait for the update to complete (with a specified timeout).

Reviewer 7: How will the timeout work and what will be the effect? I have a hard time understanding how `gardener-node-agent` could wait with a timeout when the node is rebooted because it will no longer be running. For Kubernetes minor updates, it could wait for a specified timeout, but what should it do in case of failures?
In other words, what should the effect of the timeout be?

Author: There will be two timeout inplace - 
- GNA - if update is not able to get applied in a specific timeout, it will consider it as failed update and add the label update-failed the node, so MCM knows the update failed.
- MCM - This timeout if for the case when GNA is not able to report back for ex - in case of OS failed to boot, if this timeout is reached, MCM will consider it as a failed update.

6. Once the machine has been updated, the node will be labeled by the `gardener-node-agent` with the `update-successful` label.

> 7. `machine-controller-manager` will `uncordon` any node with the `update-successful` label and remove this and all other update-related labels. When the node becomes `Ready`, it can host again workload.

Reviewer 7: When you mention a label here what does it mean? Do you mean a label key, e.g. `candidate-for-update: "true"`, or do you mean a label value, e.g. `mcm.gardener.cloud/node-state: candidate-for-update`?

It is not obvious to me if you intend to simply change the label values while moving through the states or if you want to add/remove labels in the process.

Could you please clarify?

Reviewer 3: We are using labels of the form `node.machine.sapcloud.io/ready-for-update: true` and `node.machine.sapcloud.io/update-successful: true`.

Reviewer 7: Not sure if it was just me or if it makes sense to add a short version of your comment to the document. I am fine with both.

![InplaceUpdate](assets/inplace-update.png "InPlace Update Strategy")

> #### InplaceUpdateOnLabel strategy

Reviewer 8: How will this affect `Shoot` reconciliation? Will it remain in `Processing`, waiting for the update to take place?

Reviewer 3: Updated the GEP with:
A new optional constraint will be introduced in the Shoot status to indicate that some worker pools are undergoing a manual in-place update. Although the Shoot will be marked as successfully reconciled, these worker pools might remain outdated. Furthermore, a field (can be called `pendingWorkersInPlaceUpdate`) will be introduced in the Shoot status to track worker pools awaiting in-place updates.

1. On a shoot spec change for a worker pool, `machine-controller-manager` will label all nodes of the changed worker pool with the `candidate-for-update` label. This label is used to identify nodes that require an update.
2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent it from being scaled down by `cluster-autoscaler` during the update process.

> 3. `machine-controller-manager` will wait for the `selected-for-update` label on the node. The orchestration of the update is solely the responsibility of the user. The user is free to select the nodes to be updated at will.

Reviewer 7: What will be the effect if a user does not select nodes for update for a longer period of time? If I remember correctly, Gardener reports in the shoot health conditions if a worker group is not updated to the latest operating system config after a longer period of time. Will this apply here as well? Would the time frame need to be made configurable?

Reviewer 7: From another comment, I understood that shoot reconciliation should succeed and the health condition hopefully will not fail right away, but maybe later.
Is this assumption correct?

Reviewer 3: We will add an optional constraint denoting that some updates are pending.

4. Once the `selected-for-update` label is added by the user on the node, steps 4-7 of the `Inplace` strategy will performed.

![InplaceUpdateOnLabel](assets/inplace-update-onlabel.png "InPlace Update OnLabel Strategy")

### Gardener

#### `CloudProfile` API

A new field `minSupportedVersionInPlaceUpdate` is introduced under `.spec.machineImages[]` under the CloudProfile spec. This field can be used to specify the minimum supported version from which the current machine image version can be in-place updated to.

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: CloudProfile
metadata:
  name: cloudprofile1
spec:
  type: <some-provider-name>
  providerConfig: <some-provider-specific-cloudprofile-config>
  kubernetes:
    versions:
      - version: 1.31.1
  machineImages:
    - name: gardenlinux
      versions:
        - version: 1443.8.0
          expirationDate: "2025-02-28T23:59:59Z"

>           minSupportedVersionInPlaceUpdate: 1312.3.0

Reviewer 7: Nit: The name is already quite long/verbose, but at least to me `minSupportedVersionForInPlaceUpdate` sounds better. I am no native speaker, though.

>         - version: 1443.7.0

Reviewer 7: Would this imply that I can update from `1312.3.0` or any other version below `1443.8.0` to `1443.8.0`, but I cannot use in-place update for `1443.7.0` because there the field is missing? Is there some magic going on or does this attribute be repeated on all versions?

Reviewer 3: In-place update is not possible if this field is missing.

  machineTypes:
    - name: m5.large
      cpu: "2"
      gpu: "0"
      memory: 8Gi
      usable: true
  regions:
    - name: <region-name>
```

#### `Shoot` API

> A new field `UpdateStrategy` is introduced under `spec.provider.workers[]` in the Shoot spec. This field will be passed on to the worker extension. Once set, changing the `UpdateStrategy` is prohibited through validation. Additionally, when the `InPlaceUpdate/InplaceUpdateOnLabel` update strategy is configured, skipping the worker pool's Kubernetes minor version is not allowed.

Reviewer 7: If `UpdateStrategy` is immutable then I cannot even change from `InPlaceUpdate` to `InPlaceUpdateOnLabel` (hint: there is typo, please fix) or the other way around. Is that intentional? I could imagine that I start with in-place update, but later want to have more control or the other way around.
WDYT?

Reviewer 3: It is allowed to switch between these two, will mention it explicitly.

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
metadata:
  name: crazy-botany
  namespace: garden-dev
spec:
  secretBindingName: my-provider-account
  cloudProfile:
    name: cloudprofile1
  region: europe-central-1
  provider:
    type: <some-provider-name> # {aws,azure,gcp,...}
    workers:
      - name: cpu-worker
        minimum: 5
        maximum: 5
        maxSurge: 0
        maxUnavailable: 2

>         updateStrategy: InPlaceUpdate # RollingUpdate/InPlaceUpdate/InplaceUpdateOnLabel

Reviewer 7: Should we mention what the default `updateStrategy` is?

          type: m5.large
          image:
            name: <some-image-name>
            version: <some-image-version>
            providerConfig: <some-machine-image-specific-configuration>
          architecture: <some-cpu-architecture>
  kubernetes:
    version: 1.27.3
  networking:
    type: <some-network-extension-name> # {calico,cilium}
```

#### `Worker` API

A new field `updateStrategy` is introduced under `.spec.pools[]` in the Worker API. This field will be used to set the `machineDeploymentStrategy` for the `machinev1alpha1.MachineDeployment` resource.

```yaml
apiVersion: extensions.gardener.cloud/v1alpha1
kind: Worker
metadata:
  name: worker
  namespace: shoot--foobar--aws
spec:
  type: aws
  region: eu-west-1
  secretRef:
    name: cloudprovider
    namespace: shoot--foobar--aws
  infrastructureProviderStatus: <infra-status>
  pools:
    - name: cpu-worker-1
      machineType: m4.large
      machineImage:
        name: gardenlinux
        version: 1.2.2
      minimum: 5
      maximum: 5
      maxSurge: 0
      maxUnavailable: 2
      updateStrategy: InPlaceUpdate # RollingUpdate/InPlaceUpdate/InplaceUpdateOnLabel
```

#### `OperatingSystemConfig` API

> A new field `version` is introduced in the `OperatingSystemConfig` spec. It specifies the machine image/OS version of the worker pool.

Reviewer 7: Is the `version` the desired version? How will this look like during the update? Will an updated node reference a new `version` while a non-updated one will refer to an operating system config with another `version`?

Reviewer 3: It is the desired version. The OSC will have it at all times, and after a successful update, it will be persisted as last applied OSC in the node. The node is always referring to the same OSC. The difference is calculated between the current OSC and the last applied OSC present on the node.

```yaml
apiVersion: extensions.gardener.cloud/v1alpha1
kind: OperatingSystemConfig
metadata:
  name: pool-01-original
  namespace: default
spec:
  type: <os-type>
  version: <operating-system-version>
  units:
    - ...
  files:
    - ...
```

#### Gardener Node Agent

The `gardener-node-agent` will function as the tool responsible for updating Kubernetes minor version and Machine image version dependencies on the node.

Currently, when a dependency changes in the Operating System Configuration (OSC), `gardener-node-agent` detects the difference between the current OSC and the last applied OSC, and updates any changed unit on the node. When the in-place update strategy is configured, and either the Kubernetes minor version or the machine image version is changed, `gardener-node-agent` will not immediately apply the updates to the node. Instead, it will wait for the `ready-to-update` label on the node. Once the label is present, `gardener-node-agent` will proceed with the update.

> In the case of OS updates, `gardener-node-agent` expects an "update-script" to be present in the `.spec.files[]` of the OSC at a predefined path, and the same will be executed to perform the update.

Reviewer 8: Wouldn't it be cleaner if this update script location is provided via an explicit field in `spec` instead of relying on a hard-coded path?

Reviewer 3: Agreed. We will introduce a field in the status, since extensions should be writing to this field.

> After the update, based on its success or failure, `gardener-node-agent` will add the appropriate label(`update-successful` or `update-failed`) to the node to indicate the status of the update.

Reviewer 8: When will these labels be removed?

Reviewer 5: Request that `update-successful` have a succinct, information message for the label value that is logable by the MCM and be capable of being put in the `Machine` status.  If this is not possible, please consider an the use of a separate annotation `node.machine.sapcloud.io/update-message`.

Author: MCM removes all the labels after the update is successful and the machine is moved to the new machine set.

There are two ways to achieve this:

1. **Owner Reference Update**: Move the old `Machine` to the new `MachineSet` by updating the owner reference. The downside of this method is that the `Machine` name will stay the same through updates, which breaks the current naming convention of including the `MachineSet` name as a prefix in `Machine` name.
2. **Shallow Delete**: Instead of directly moving a `Machine` from one `MachineSet` to another, first perform a shallow delete of the old `Machine` object (without deleting the underlying physical/virtual machine), then create a new `Machine` object in the new `MachineSet` (for the same physical/virtual machine).

> [!NOTE]
> At the time of writing this GEP, it is undecided which method to proceed with. The first method has been evaluated, but the feasibility of the second method is yet to be assessed. If feasible, the second method is preferred.

#### `MachineDeployment` API

> A new strategy type `InPlaceUpdate` and a new field, `inPlaceUpdate`, have been introduced under the `.spec.strategy` field. `inPlaceUpdate` can be configured when the selected strategy type is `InPlaceUpdate`.

Reviewer 7: Nit: In previous sections you write that ... "is introduced", but here you write ... "have been introduced". Maybe you want to be consistent with regards to time.

```yaml
apiVersion: machine.sapcloud.io/v1alpha1
kind: MachineDeployment
metadata:
  name: <name>
  namespace: <shoot-namespace>
spec:
  replicas: 1
  selector:
    matchLabels:
      name: <label>
  strategy:
    inPlaceUpdate:
      maxUnavailable: 1
      maxSurge: 0
      onLabel: false # true/false
    type: InPlaceUpdate # RollingUpdate/InPlaceUpdate/Recreate
  template: {}
    spec: {}
status: {}
```

`maxUnavailable`: Specifies the maximum number of unavailable machines during the update process.

> `maxSurge`: Specifies the maximum number of machines that can be scheduled above the desired number of machines. Set `maxSurge` to zero if you don't want any new machines/nodes to be provisioned during the update. You can set it to a value greater than zero if you don't want unavailable nodes during the update; in that case, first new machines/nodes will be created with the updated configuration and then the old machines/nodes will undergo in-place updates.

Reviewer 7: Setting `maxSurge` to something greater than zero may lead to new machines being created and hence having more nodes than before the update. Therefore, some nodes may be scaled down later on.
Would this not somewhat reduce the effect of the in-place update as you may "loose" old machines over time? I am all for the flexibility, but I am a bit confused if this is something anybody would actually want.
WDYT?

Reviewer 2: Just a comment: Whether `maxSurge` or out-scaling through the cluster-autoscaler, in-place updates with a drain (in contrast to the non-drain option that was rejected until now) may easily result in a situation where new nodes are necessary, because we remove capacity from the cluster during the drain/reboot and the pods must go somewhere then (unless they are node-sticky, but that will probably be the minority of the cases).

Of course, for manual bare metal clusters, `maxSurge` cannot be anything else than 0 and `minimum` must be equal to `maximum` to disable the cluster-autoscaler. Of course, we haven't yet defined how manual bare metal clusters will look like at spec level, but I imagine they must somewhat look like that. :-)

Reviewer 2: In order to continue and close this discussion and in general, we should sharpen the above sentence, e.g.:

```suggestion
`maxSurge`: Specifies the maximum number of machines that can be scheduled above the desired number of machines. Set `maxSurge` to zero if you don't want any new machines/nodes to be provisioned during the update. You can set it to a value greater than zero if you like to avoid pending pods because of lack of capacity during the update; in that case, first new machines/nodes will be created with the updated configuration and then the old machines/nodes will undergo in-place updates.
```

Is that how I should imagine in-place updates with `maxUnavailable` > `0` and `maxSurge` > `0`, probably `maxUnavailable` == `maxSurge`? We do not usually have this case with the old rolling updates, but with in-place updates, existing nodes will become unavailable during their in-place update as we decided to drain them, but at the same time, most users will prefer to provision additional nodes to take on the drained pods. Of course, one could also rely on CA to provision the extra capacity, but like with the rolling updates, if you already know/anticipate the need for nodes because of the update, you may want to have them already ready.

Only the "real" (=manually provisioned) bare metal use case will set `maxSurge` to `0`, but with virtual machines, it will definitely be safer and faster to make sure to have extra capacity ready for the drained pods.

Author: Yes. In case of InPlace update with both `maxSurge` and `maxUnavaiable`, the new machines will get created first respecting the `maxSurge` number and then `MCM` logic will either perform InPlace updates on the old machines or delete them based on the need and respecting `maxUnavaiable`.

`onLabel`: If the update strategy is `InPlaceUpdateOnLabel` in the worker spec, the worker controller will set this field to true.

### Dependency Watchdog

Currently, the [`dependency-watchdog`](https://github.com/gardener/dependency-watchdog) monitors the leases of the nodes in the shoot cluster. If more nodes than a configured threshold are unhealthy, it scales down the `machine-controller-manager`, `kube-controller-manager`, and `cluster-autoscaler` to prevent a meltdown. During an ongoing in-place update, DWD can make use of the labels added by `machine-controller-manager` during the `in-place` update to exclude nodes undergoing updates from its health check.

### Extensions

#### OS extensions

> The OS extension that wants to support in-place updates should add an "update-script" to the `.spec.files[]` of the OSC with a pre-defined path that will be executed by the `gardener-node-agent` to perform the update. This script should trigger the OS-provided update tool/utility to update the OS.

Reviewer 8: See above

### Failures and Recovery Strategy

In-place updates may fail because of several factors ranging from network issues to compatibility problems. This section outlines the failure scenarios, error classifications, and the specific roles `gardener-node-agent` and `machine-controller-manager` play in failure handling and recovery to ensure minimal disruption.

#### Types of Failures During Update

##### Drain Timeout/Failure

- A node is marked as `ready-for-update` only after a successful drain.
- We use the existing [drain logic](https://github.com/gardener/machine-controller-manager/tree/v0.54.0/pkg/util/provider/drain) provided by the `machine-controller-manager`.

> - If a drain operation times out or fails, the node will be forcefully drained.

Reviewer 8: What does "forcefully drain" mean? Is this just a regular pod deletion?

Reviewer 3: Yes. MCM deletes all the pods without any grace period.

Reviewer 2: Is this discussion now closed? @rfranzke is on parental leave.

Reviewer 3: Yes, I think we can.

##### Kubernetes Update Failures

> - **Rollback on Failure:** If a Kubernetes version update fails, `gardener-node-agent` will roll back the `kubelet` to the version specified in the last applied `OperatingSystemConfig`. This ensures that the node remains operational.

Reviewer 8: How will this work? The OSC secret got already updated to the new kubelet config/files, etc. How can it roll back/how will this be orchestrated? Will `gardenlet` update the OSC secret again to the old version? But what happens to the version in the `Shoot` spec then which gets inconsistent?

Reviewer 3: We thought about it and it might not be a very good idea to provide this feature. OS updates needs rollback because we can no longer communicate with the node without a functioning OS. This is not the case with the kubelet. Will update the doc.

  - **Update Halt on Failure:** `gardener-node-agent` will label the node with an `update-failed` label. This label signals `machine-controller-manager` to stop further updates.

##### OS Update Failures

- **Before rebooting for update** - Update cannot start if the machine image version is unavailable, network issues prevent the image pull, or disk storage is insufficient (although proper garbage collection should prevent this); classified as retriable or non-retriable.
- **Failed to boot the new OS** - If the node fails to boot into the new OS version and reverts to the previous version, `gardener-node-agent` will annotate it to mark an update attempt, but the root cause may remain unlogged as filesystems mount later in the booting process; classified as non-retriable.
- **Total failure** - In cases where the system fails to boot into either of the OS versions, the update attempt will timeout and be marked as failed; classified as non-retriable.

![FailureStrategy](assets/inplace-failurestrategy.png "InPlace Failure State Diagram")

#### Roles in Failure Handling

**`gardener-node-agent`**

> `gardener-node-agent` is responsible for initiating updates, with specific timeouts set for Kubernetes and machine image updates. When an error occurs it will be classified as follows:

Reviewer 7: This looks duplicative in the rendered view, not sure how you expect it to look. Maybe we should remove one `gardener-node-agent`. Same is valid for `machine-controller-manager` below.

- **Retriable Error**: Temporary errors, such as network interruptions, prompt `gardener-node-agent` to retry the update after a delay. Retriable error will also get classified as non-retriable error after a certain number of retries and `gardener-node-agent` labels the node with an `update-failed` label.

> - **Non-Retriable Error**: For critical issues, such as incompatible OS versions, `gardener-node-agent` labels the node with an `update-failed` label.

Reviewer 7: If we want to make `gardener-node-agent` classify operating system issues, we assume that `gardener-node-agent` will get running after big issues, e.g. a corrupt operating system image. Is that assumption valid? Should this be stated explicitly?

Author: Only the error returned when the OS update script is not executed succesfully will get classified, like version not found, unable to pull the new image and so and so.
For the case where the OS fails to boot/crashes, we will anyway not get the errors of why the OS update failed, and GNA will not be able to come up, in this case, MCM will timeout, and consider it a failed update.

**`machine-controller-manager`**
`machine-controller-manager` prepares nodes for updates by adding the `ready-to-update` label and waits for `gardener-node-agent` to perform the update. `machine-controller-manager` monitors for status labels applied by `gardener-node-agent`:

> - **Timeout Handling**: After the `machine-controller-manager` adds the `ready-for-update` label to a node, it waits for the `gardener-node-agent` to complete the update within the timeout configured in `machine-controller-manager`. If `gardener-node-agent` does not complete the update within this timeout, `machine-controller-manager` will retry the rollout after a specified delay. The timeout on the `machine-controller-manager` side is essential for handling cases where `gardener-node-agent` fails to report any status, such as during a total failure of the OS update. If the update is unsuccessful after a certain number of retries, `machine-controller-manager` will add the `update-failed` label to the node.

Reviewer 7: How do we ensure that this will not be a race? If both `gardener-node-agent` and `machine-controller-manager` may set the update label what happens if we run close to the timeout and they overwrite each others value? Any thoughts?

Reviewer 3: We are using labels of the form `node.machine.sapcloud.io/ready-for-update: true` and `node.machine.sapcloud.io/update-successful: true`. `gardener-node-agent` and MCM will be working on different labels and they should not be writing to the same label at any point.

- **Failed Update Handling**: If a node is labeled with `update-failed`, `machine-controller-manager` stops further updates. The machine remains with the old `MachineSet`, and no additional nodes in that `MachineDeployment` will be updated until the node with `update-failed` label is fixed. The failure is propagated to the shoot status, notifying the operator or user, who may need to investigate the issue manually.
  To resolve the issue, the error must be fixed manually, and the `update-failed` label should be removed from the node. This allows `machine-controller-manager` to proceed with the update process.

> [!NOTE]
> All mentioned timeouts and retries are subject to change, with exact values to be decided during implementation.

### Additional Validations

> In addition to Kubernetes or Machine image updates, several other factors can trigger the rolling updates of nodes in the shoot clusters. These include:

Reviewer 7: Should we link to the documentation here where we list all cases?

1. Modification of the following fields in the Shoot specification:
   - `.spec.provider.workers[].machine.type`
   - `.spec.provider.workers[].volume`
   - `.spec.provider.workers[].cri`
   - `.spec.systemComponents.nodeLocalDNS`

> 2. Rotation of the `CA` or `ServiceAccount` signing key.

Reviewer 8: How will you facilitate the needed updates to kubelet to get the new credentials?

3. Changes in the `kubelet` configuration (for eg., kubeReserved).

In-place updates will support (2) and (3) but will disallow (1) with validations in the Shoot API. This ensures that certain critical changes, such as machine type or volume adjustments, follow the traditional rolling update process in which nodes are replaced.

## Future Work

- Support node updates without graceful termination of the pods (via node drain) scheduled onto the node.
- Allow to switch the update strategy from `RollingUpdate` to `InplaceUpdate`/`InplaceUpdateOnLabel` or vice-versa.

## Alternatives

### Using Only Gardener Node Agent

Instead of using both the `machine-controller-manager` and the `gardener-node-agent` for in-place updates, use only `gardener-node-agent`. However, this method has the following drawbacks:

> - `machine-controller-manager` natively provides rolling update logic through `Machine Deployment` and `Machine Set`, which helps track the progress of updated and pending nodes. Additionally, `machine-controller-manager` includes logic to select machines based on `maxUnavailable/maxSurge`, ensuring machines are rolled in a controlled manner. If only `gardener-node-agent` is used, new logic would need to be implemented to track node update progress and to manage the number of nodes updated at a time before proceeding to the next set.

Reviewer 8: ```suggestion
- `machine-controller-manager` natively provides rolling update logic through `Machine Deployment` and `Machine Set`, which helps track the progress of updated and pending nodes. Additionally, `machine-controller-manager` includes logic to select machines based on `maxUnavailable/maxSurge`, ensuring machines are rolled in a controlled manner. If only `gardener-node-agent` was used, new logic would need to be implemented to track node update progress and to manage the number of nodes updated at a time before proceeding to the next set.
```

    - [Dependency Watchdog](#dependency-watchdog)
    - [Extensions](#extensions)
      - [OS extensions](#os-extensions)
    - [Worker pool hash calculations](#worker-pool-hash-calculations)
    - [Failures and Recovery Strategy](#failures-and-recovery-strategy)
      - [Types of Failures During Update](#types-of-failures-during-update)
        - [Drain Timeout/Failure](#drain-timeoutfailure)
        - [Kubernetes Update Failures](#kubernetes-update-failures)
        - [OS Update Failures](#os-update-failures)
      - [Roles in Failure Handling](#roles-in-failure-handling)
  - [Future Work](#future-work)
  - [Alternatives](#alternatives)
    - [Using Only Gardener Node Agent](#using-only-gardener-node-agent)
    - [Using the same name for `MachineClass`](#using-the-same-name-for-machineclass)

## Summary

Gardener should support new update strategies for nodes that do not require the deletion and recreation of the nodes. These strategies aim to minimize the overhead traditionally associated with node replacement and offer an alternative approach to updates, which is particularly important for physical machines or bare-metal nodes.

> The in-place update strategy is not intended to replace the current rolling update strategy provided by the [`machine-controller-manager`](https://github.com/gardener/machine-controller-manager). Instead, it is designed for specific scenarios where physical machines or bare metal nodes, whether manually joined or programmatically provisioned, cannot be easily rolled. Additionally, this update strategy can be useful in scenarios where rolling machines are undesirable, such as when dealing with large machines or limited resources on the cloud provider. For all other scenarios where preserving nodes is not critical, the rolling update strategy remains the preferred approach.

Reviewer 2: Why do you write "For all other scenarios where preserving nodes is not critical, the rolling update strategy remains the preferred approach"? Why do you recommend rolling the nodes?

It takes longer, always bears the risk of not getting the hardware, etc., so where do you see an advantage in replacing instead of updating the nodes? If updating works (and that's the goal), then it will become my go-to choice and I will not roll anymore. Why would I?

```suggestion
The in-place update strategy is not intended to replace the current rolling update strategy provided by the [`machine-controller-manager`](https://github.com/gardener/machine-controller-manager). Instead, it offers an alternative that is especially useful in situations where physical machines or bare metal nodes are involved (manually joined or programmatically provisioned) that cannot be rolled easily. Additionally, this update strategy can be useful in scenarios where rolling machines is undesirable, such as when dealing with large machines or limited resources on the cloud provider.
```

Reviewer: In order to be able to update the BIOS or the Firmware of the server, the server must be powered off/on after the new BIOS/Firmware was applied. This is required for security reasons on a regular basis, but if you have the servers occupied all year long they will never receive critical updates as such.

Reviewer 2: I understand @majst01, but isn't it a problem to rely on such "by-chance updates" (just because something changes that throws away the machine, the security problem is also gone - but only as a side effect).

Machines should be updated/rebooted, depending on the severity of the vulnerabilities/fixes, e.g. reboot within 1 day after resolving a vulnerability of score >9, rebooting after 7 days after resolving a vulnerability of score >8, etc. or whatever the strategy of the operating company is.

Besides, in-place updates don't conflict with rebooting. In-place only means, that machines are not "released", which isn't possible with bare metal nodes/for edge use cases in the future. We have no choice there. But rebooting should always be possible and in-place isn't ruling that out. In fact, it's part of some of the update scenarios, e.g. OS updates as discussed here will reboot the machine.

Reviewer: Yes, but a reboot will not power cycle a machine which is required for most firmware updates. And as side note, updating the firmware is often a risky, but still required, action which in some cases brick the machine, so this will completely bring down the node if done with the in-place strategy.

Reviewer 2: Ah, I should have paid more attention to BIOS/firmware in your first comment, I am sorry @majst01. Let's loop in @afritzler as well.

In this case, from the point of view of the Gardener/Kubernetes node, the node doesn't probably need to know, because it doesn't need to change (nothing on disk needs to change), i.e. it goes down, something is done with it (not unlike some other hardware maintenance), and it comes up again as the very same node. Would you agree?

So, this scenario can/must be orchestrated completely from the outside (as any other hardware maintenance), e.g. cordon/drain, perform said BIOS/firmware upgrade, power up again. It's not really an in-place update from the PoV of Gardener, is it? Gardener doesn't need to know and doesn't need to play along or do something here.

Of course, in case you have "some metal API" that helps and deals with management of bare metal infrastructure, it would be great if MCM could help orchestrate such things, but that's not what this GEP is about.

  - `.spec.provider.workers[].machine.image.version`
  - `.status.credentials.rotation.{certificateAuthorities,serviceAccountKey}.lastInitiationTime`
  - `.spec.kubernetes.kubelet.{kubeReserved,systemReserved,evictionHard,cpuManagerPolicy}` or `.spec.provider.workers[].kubernetes.kubelet.{kubeReserved,systemReserved,evictionHard,cpuManagerPolicy}`
- Disallow any other update triggers with validation when the strategy is set to an in-place strategy.
- Provide two strategies for in-place - `AutoInPlaceUpdate` and `ManualInPlaceUpdate`. The former will be orchestrated by the `machine-controller-manager` and the latter by the users themselves.
- Switching between `AutoInPlaceUpdate` and `ManualInPlaceUpdate` strategies.

### Non-Goals

- Provide a way to schedule the pods on the same node after the in-place update.
- Downgrading of Kubernetes/machine image version after successful update.
- Switching between in-place and rolling update strategies.
- Provide an update strategy without draining the node.
- Provide in-place updates only for the Kubernetes version in worker pools where the OS does not support in-place updates.

> - Provide in-place updates for any rolling update trigger other than the ones mentioned in the [Goals](#goals).

Reviewer 2: The actual goal (of the ticket/issue; the very first sentence) was to provide in-place updates for everything that is not related to the node hardware (blacklist-approach), but this goal/non-goal turns it into something different by listing for which changes in-place updates are available (whitelist-approach).

That was not the idea. Can you please list, what besides changing the hardware (machine, attached volumes, etc.) is not covered by in-place updates? The bare metal scenario will not work with a 99% solution. We can discuss that very special features are not supported, but the premise should always be: aside from changes to the actual node hardware, all spec changes can be executed in-place.

Reviewer 3: Rewrote the goals. PTAL.

## Proposal

In short, Gardener should support new update strategies that do not require deleting and recreating the node when:

1. Kubernetes minor version is updated
2. Machine image version is updated
3. Certificate Authority (CA) rotation or ServiceAccount signing key rotation is triggered

> 4. Relevant Kubelet configuration is modified.

Reviewer 2: See above, you turned the blacklist definition into a whitelist definition, no longer respecting the very first sentence of the original ticket/issue which is:

> We need to offer, for **all** spec changes, a way to update cluster nodes in-place (all except for the ones where we have no other technical alternative, e.g. switching out physical machines will always require node rolling).

### Approach

The `machine-controller-manager` and `gardener-node-agent` components will be used to perform in-place updates. `machine-controller-manager` will provide an interface that prepares the machine/node for an update. This interface can be used by other components to determine if a node is ready for an update, allowing the other component to perform the update on the node and then notify the `machine-controller-manager` that the node is successfully updated. _In the case of Gardener, this component will be the `gardener-node-agent`._

This design was chosen to leverage the `machine-controller-manager`'s node rollout orchestration functionality while ensuring that it remains a standalone component.

### Prerequisites

> The only prerequisite is that the machine image/OS on the node must support in-place updates and provide a tool or utility to initiate the OS update and boot into the new version. This update tool/utility should meet the following requirements:

Reviewer 2: This prerequisite is a specific one only if the OS is changed, so let's add that note:

```suggestion
To update the OS in-place, the only prerequisite is that the machine image/OS on the node must support in-place updates and provide a tool or utility to initiate the OS update and boot into the new version. This update tool/utility should meet the following requirements:
```

1. **Failure Recovery**: In case of a failure, the OS should be able to revert to the previous version.
2. **Retry Configuration**: Support for configuring the number of update retries.

> 3. **Registry Configuration**: Ability to specify the registry from which to pull the OS image.

Reviewer 2: Just saying: That (an OCI registry) is already a very specific prerequisite that only applies to Garden Linux.

Reviewer 3: Rewrote the step.

6. Once the machine is updated, `gardener-node-agent` will delete all the pods in the node so that they will get recreated and the node will be labelled by the `gardener-node-agent` with the `node.machine.sapcloud.io/update-successful` label.
7. `machine-controller-manager` will `uncordon` any node with the `node.machine.sapcloud.io/update-successful` label and remove this and all other update-related labels. Once the node becomes `Ready`, it can host workload again.

![AutoInPlaceUpdate](assets/31-auto-inplace-update.png "Auto In Place Update Strategy")

#### `ManualInPlaceUpdate` strategy

1. On a shoot spec change for a worker pool, `machine-controller-manager` will label all nodes of the changed worker pool with the `node.machine.sapcloud.io/candidate-for-update` label. This label is used to identify nodes that require an update.
2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent them from being scaled down by `cluster-autoscaler` during the update process.

> 3. `machine-controller-manager` will wait for the `node.machine.sapcloud.io/selected-for-update` label on the node. The orchestration of the update is solely the responsibility of the user. The user is free to select the nodes to be updated at will.

Reviewer 2: Will we have an admission plugin that prevents users from labelling more than `maxUnavailable` nodes with `selected-for-update` (see meaning of this field for `AutoInPlaceUpdate` above; step 3)?

Reviewer 3: No, this is up to the user to decide. All the labelled nodes will undergo an update and `maxUnavailable` has no role here.

Reviewer 2: Hmm... a bit strange as this means, the user can shoot himself in the foot this way. But then, what's the meaning of `maxUnavailable` if we don't care at this step? You have not written, that `maxUnavailable` is irrelevant for manually triggered updates, so people might/will set it, but it means nothing (in this case anymore)?

Author: Putting `maxUnavailable` limit seems a bit strange as one selecting this strategy should be very well aware of what they are doing, we will add a line that `maxUnavailable` is irrelevant for manually triggered updates.

4. Once the `node.machine.sapcloud.io/selected-for-update` label is added by the user, steps 4-7 of the `AutoInPlaceUpdate` strategy will be performed.

![ManualInPlaceUpdate](assets/31-manual-inplace-update.png "Manual In Place Update Strategy")

A new optional [constraint](https://github.com/gardener/gardener/blob/master/docs/usage/shoot/shoot_status.md#constraints) will be introduced in the Shoot status to indicate that some worker pools are undergoing a manual in-place update. Although the Shoot will be marked as successfully reconciled, these worker pools might remain outdated. Furthermore, a field (can be called `pendingWorkersInPlaceUpdate`) will be introduced in the Shoot status to track worker pools awaiting in-place updates.

> Subsequent updates to the worker pool will be blocked by validation if an in-place update is already in progress. This ensures that worker pools do not skip intermediate Kubernetes minor versions or machine image versions. However, if an in-place update fails and a fix—such as a patch to the current updated minor version—is required, it will be allowed if the Shoot has the `shoot.gardener.cloud/force-in-place-update: true` annotation.

Reviewer 2: ```suggestion
Subsequent updates to the worker pool will be blocked by validation if an in-place update is already in progress. This ensures that worker pools do not skip intermediate Kubernetes minor versions or machine image versions. However, if an in-place update fails and a fix - such as a patch to the current updated minor version - is required, it will be allowed, if the Shoot has the `shoot.gardener.cloud/force-in-place-update: true` annotation.
```

### Gardener

#### `CloudProfile` API

A new field `minVersionForInPlaceUpdate` is introduced under `.spec.machineImages[].versions[]` in the `CloudProfile`. This field specifies the minimum machine image version eligible for direct in-place upgrades to a newer version. If the current version is lower than `minVersionForInPlaceUpdate`, the update will be disallowed by validation.

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: CloudProfile
metadata:
  name: cloudprofile1
spec:
  type: <some-provider-name>
  providerConfig: <some-provider-specific-cloudprofile-config>
  kubernetes:
    versions:
      - version: 1.31.1
  machineImages:
    - name: gardenlinux
      versions:
        - version: 1443.8.0
          expirationDate: "2025-02-28T23:59:59Z"
          minVersionForInPlaceUpdate: 1312.3.0
        - version: 1443.7.0 # if minVersionForInPlaceUpdate is not specified, in-place update cannot be performed on this version
  machineTypes:
    - name: m5.large
      cpu: "2"
      gpu: "0"
      memory: 8Gi
      usable: true
  regions:
    - name: <region-name>
```

#### `Shoot` API

> A new field `updateStrategy` is introduced under `spec.provider.workers[]` in the Shoot spec. This field will be passed on to the worker extension. Once set, changing the `updateStrategy` between in-place and replace strategies is prohibited through validation. Additionally, when `AutoInPlaceUpdate` or `ManualInPlaceUpdate` is configured as an update strategy, skipping an intermediate Kubernetes minor version is no longer allowed, because of potential local changes that are then not carried out anymore when skipping versions (only allowed with `AutoReplace` as the machines/nodes are created from scratch in this case).

Reviewer 2: What kind of "local changes" are meant here? Are those related to Gardener (then why, because we are in control) or the Kubelet or something else?

E.g. https://kubernetes.io/releases/version-skew-policy/#kubelet-1 doesn't specify this restriction, but that's of course not proving that it's possible.

Reviewer 2: While you changed the text, the meaning has not changed also in the new text:

> skipping an intermediate Kubernetes minor version is no longer allowed, because of potential breaking changes that are then not carried out anymore when skipping versions

So, I am wondering (see above), where is this restriction coming from?

Reviewer 3: We are not entirely certain about this, but the restriction for the "rolling of nodes" scenario was removed with #9184:
> When upgrading the worker group, all nodes are rolled anyway, hence there is no node identity that needs to be preserved or might undergo potential conversions.

and this restriction has been present from the very first version of gardener: https://github.com/gardener/gardener/commit/77abd21f099596203c0d6bd87c4197a5860735d7. 

We are unsure if skipping minor versions now could potentially cause issues .

Reviewer 2: OK, so we are not really sure and it's a safety measure. Maybe it would be possible, but it's not that relevant and at the same time risky, so we rather would not want to allow it. Got it.

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: Shoot
metadata:
  name: crazy-botany
  namespace: garden-dev
spec:
  secretBindingName: my-provider-account
  cloudProfile:
    name: cloudprofile1
  region: europe-central-1
  provider:
    type: <some-provider-name> # {aws,azure,gcp,...}
    workers:
      - name: cpu-worker
        minimum: 5
        maximum: 5
        maxSurge: 0
        maxUnavailable: 2

>         updateStrategy: AutoInPlaceUpdate # AutoReplaceUpdate/AutoInPlaceUpdate/ManualInPlaceUpdate, defaulted to  AutoReplaceUpdate

Reviewer 2: ```suggestion
        updateStrategy: AutoInPlaceUpdate # AutoReplaceUpdate/AutoInPlaceUpdate/ManualInPlaceUpdate, defaulted for now to AutoReplaceUpdate
```

2. **Shallow Delete**: Instead of directly moving a `Machine` from one `MachineSet` to another, first perform a shallow delete of the old `Machine` object (without deleting the underlying physical/virtual machine), then create a new `Machine` object in the new `MachineSet` (for the same physical/virtual machine).

> [!NOTE]
> At the time of writing this GEP, it is undecided which method to proceed with. The first method has been evaluated, but the feasibility of the second method is yet to be assessed. If feasible, the second method is preferred.

#### `MachineDeployment` API

A new strategy type `InPlaceUpdate` and a new field, `inPlaceUpdate`, are introduced under the `.spec.strategy` field. `inPlaceUpdate` can be configured when the selected strategy type is `InPlaceUpdate`.

```yaml
apiVersion: machine.sapcloud.io/v1alpha1
kind: MachineDeployment
metadata:
  name: <name>
  namespace: <shoot-namespace>
spec:
  replicas: 1
  selector:
    matchLabels:
      name: <label>
  strategy:
    inPlaceUpdate:
      maxUnavailable: 1
      maxSurge: 0
      manualUpdate: false # true/false, defaulted to false

>     type: InPlaceUpdate # RollingUpdate/InPlaceUpdate/Recreate, defaulted to RollingUpdate

Reviewer 2: I think, today we only have `RollingUpdate`. `InPlaceUpdate` is the new one. What is `Recreate`?

Reviewer 3: It's part of the `MachineDeployment` API and we don't plan to change/remove it with this GEP.

Reviewer 2: And what's the meaning, i.e. difference between `RollingUpdate` and `Recreate`?

Author: In case of `Recreate` all the old Machines are deleted at once and then the new machines get created.

Reviewer 3: The `Recreate` strategy in the MCD is not configurable via the Shoot/gardener as of today.

Reviewer 2: Yes, I know, but what does it mean? I am just trying to understand the mapping here and don't know what that MCM strategy means.

Author: > Yes, I know, but what does it mean? I am just trying to understand the mapping here and don't know what that MCM strategy means.

I am a bit confused, we kept `Recreate` because it can be configured as of today in Machine Deployment, though there is no way to do it through the shoot.
There is no mapping as such for Recreate, we just put all the available values that can be configured.

### Failures and Recovery Strategy

In-place updates may fail because of several factors ranging from network issues to compatibility problems. This section outlines the failure scenarios, error classifications, and the specific roles `gardener-node-agent` and `machine-controller-manager` play in failure handling and recovery to ensure minimal disruption. The error will be reported in `node.machine.sapcloud.io/update-failure-message` annotation on the node which can be used subsequently by the Shoot care controller.

#### Types of Failures During Update

##### Drain Timeout/Failure

- A node is marked as `node.machine.sapcloud.io/ready-for-update` only after a successful drain.
- We use the existing [drain logic](https://github.com/gardener/machine-controller-manager/tree/v0.54.0/pkg/util/provider/drain) provided by the `machine-controller-manager`.
- If a drain operation times out or fails, the node will be forcefully drained.

##### Kubernetes Update Failures

> - **Update Halt on Failure:** `gardener-node-agent` will label the failed node with an `node.machine.sapcloud.io/update-failed` label. This label signals `machine-controller-manager` to stop further updates when the number of unhealthy machines reaches configured `maxUnavailable`.

Reviewer 2: Hmm... so what is the exact meaning of `maxUnavailable` now for in-place updates? See also question above, how many nodes the user may label with `selected-for-update` with `ManualInPlaceUpdate`.

I guess, the sum total of nodes with either `selected-for-update`, `ready-for-update`, **and** `update-failed` should never go beyond `maxUnavailable`? If that's the case, we should say so in the GEP.

So, in natural language, `maxUnavailable` expresses how many nodes can be updated in parallel **and** also expresses how many nodes are fubar/broken at most after an update?

Reviewer 4: `maxUnavailable` should semantically mean that during an update of a cluster, the max machines that are unavailable should not be more than this value. The reason could be anything.

Ideally `selected-for-update` should be marked on a node after analysing the current `updated-failed` and nodes which have already been marked with `selected-for-update` or `ready-for-update` labels.

Author: All the machines undergoing in-place updates are accounted as unavailable machines. A machine is considered undergoing an in-place update as soon as it gets labeled with `selected-for-update` label.
For more clarification we will add one line under MCM API `maxUnavaiable` field stating - `Any machine undergoing an in-place update or with an update failed will be counted as unavailable.`

Reviewer 2: > For more clarification we will add one line under MCM API `maxUnavaiable` field stating - Any machine undergoing an in-place update or with an update failed will be counted as unavailable.

Thanks.

So my suggestion, "the sum total of nodes with either `selected-for-update`, `ready-for-update`, and `update-failed` will never go beyond maxUnavailable" is wrong because in the manual scenario, a user can label any number of nodes with `selected-for-update` because we have no admission plugin that would prevent that, but that's already a thread of discussion above, so I will resolve this discussion.

![FailureStrategy](assets/31-inplace-failurestrategy.png "InPlace Failure State Diagram")

#### Roles in Failure Handling

**`gardener-node-agent`**:

`gardener-node-agent` is responsible for initiating updates, with specific timeouts set for Kubernetes and machine image updates. When an error occurs it will be classified as follows:

- **Retriable Error**: Temporary errors, such as network interruptions, prompt `gardener-node-agent` to retry the update after a delay. Retriable error will also get classified as non-retriable error after a certain number of retries and `gardener-node-agent` labels the node with an `node.machine.sapcloud.io/update-failed` label.
- **Non-Retriable Error**: For critical issues, such as incompatible OS versions, `gardener-node-agent` labels the node with a `node.machine.sapcloud.io/update-failed` label.

**`machine-controller-manager`**:

`machine-controller-manager` prepares nodes for updates by adding the `node.machine.sapcloud.io/ready-to-update` label and waits for `gardener-node-agent` to perform the update. `machine-controller-manager` monitors for status labels applied by `gardener-node-agent`:

> - **Timeout Handling**: After the `machine-controller-manager` adds the `node.machine.sapcloud.io/ready-for-update` label to a node, it waits for the `gardener-node-agent` to complete the update within the timeout configured in `machine-controller-manager`. If `gardener-node-agent` does not complete the update within this timeout, `machine-controller-manager` will retry the rollout after a specified delay. The timeout on the `machine-controller-manager` side is essential for handling cases where `gardener-node-agent` fails to report any status, such as during a total failure of the OS update. If the update is unsuccessful after a certain number of retries, `machine-controller-manager` will add the `node.machine.sapcloud.io/update-failed` label to the node.

Reviewer 2: What is the exact meaning of "`machine-controller-manager` will retry the rollout after a specified delay"? I mean, the node is already labelled with `ready-for-update`, so is it taking away that label and re-applying it later (after the specified delay)? I ask, because at that point in time, MCM is no active actor anymore and became merely an observing watcher.

Reviewer 4: I agree, once MCM adds the label `ready-for-update` on a node, then it has to wait for a signal from the node agent  marking the completion of update (successfully or unsuccessfully). In that context, the following line needs to be revisited.
>If `gardener-node-agent` does not complete the update within this timeout, `machine-controller-manager` will retry the rollout after a specified delay.

Currently there is a clear demarcation of responsibility - MCM orchestrates by selecting one or more nodes that needs an in-place update and marks them with specific labels. If that number has already reached the `maxUnavailable` then it pauses and waits for response from node-agent to notify (in this case the notification will be by adding a label as an example). If there is no notification from the node-agent, then AFAIK, currently it does not have a way to identify what transpired.

If you wish to now add a timeout, then you can potentially have 2 actors try to update the status of an update which could potentially lead to lost updates issue.

Author: > The timeout on the `machine-controller-manager` side is essential for handling cases where `gardener-node-agent` fails to report any status, such as during a total failure of the OS update.

MCM needs separate timeouts for the above mentioned reason.

> If `gardener-node-agent` does not complete the update within this timeout, `machine-controller-manager` will retry the rollout after a specified delay.

Here we meant that the number of retries specifies how many times the `machine-controller-manager` will attempt the timeout period before concluding that the `gardener-node-agent` is unable to complete the update and marking the process as failed.
 
> If `gardener-node-agent` does not complete the update within this timeout, `machine-controller-manager` will retry the rollout after a specified delay. 

We will remove this line as retires are not needed and one timeout should be enough on MCM end.

> - **Failed Update Handling**: If a node is labeled with `node.machine.sapcloud.io/update-failed`, `machine-controller-manager` stops further updates. The machine remains with the old `MachineSet`, and no additional nodes in that `MachineDeployment` will be updated until the node with the `node.machine.sapcloud.io/update-failed` label is fixed. The failure is propagated to the shoot status, notifying the operator or user, who may need to investigate the issue manually.

Reviewer 2: Hmm... this seems to contradict the statement made in [line 381](https://github.com/gardener/gardener/pull/10828/files#diff-13a13e701eb5f7f1198c82b8d6c7a8b58902eac36d631b1f828d65a2a015093eR381):

> This label signals `machine-controller-manager` to stop further updates **when the number of unhealthy machines reaches configured `maxUnavailable`**.

So, what is it, `1` or `maxUnavailable`? `maxUnavailable` is defined per worker pool and there is a 1:1 relationship to the machine deployment, so both statements cannot be both true, I would think.

Reviewer 3: Rewrote it.

Reviewer 2: Thanks!

  To resolve the issue, the error must be fixed manually, and the `node.machine.sapcloud.io/update-failed` label should be removed from the node. This allows the `machine-controller-manager` to proceed with the update process.

> [!NOTE]
> All mentioned timeouts and retries are subject to change, with exact values to be decided during implementation.

## Future Work

> - Support node updates without graceful termination of the pods (via node drain) scheduled onto the node.

Reviewer 2: I would think, the below suggestion is more accurate as some pods may actually be gracefully terminated (the individual grace periods are curtailed by the grace period for the node, but why wouldn't we try, on best effort, to gracefully terminate all pods with a given node-global grace period first - there is no harm in this).

At at any rate, the main point is not the termination, but not doing the drain, which effectively drops the pod entirely (a new pod must be created by a replication controller after the drain, with all consequences this bears):

```suggestion
- Support node updates without draining the scheduled pods first/at all.
```

- Allow to switch the update strategy from `AutoReplaceUpdate` to `AutoInPlaceUpdate`/`ManualInPlaceUpdate` or vice-versa.
- Evaluate whether nodes can consistently be updated in-place for supported update triggers, eliminating the need to offer a choice between in-place and rolling updates.

## Alternatives

### Using Only Gardener Node Agent

Instead of using both the `machine-controller-manager` and the `gardener-node-agent` for in-place updates, use only `gardener-node-agent`. However, this method has the following drawbacks:

- `machine-controller-manager` natively provides rolling update logic through `MachineDeployment` and `MachineSet`, which helps track the progress of updated and pending nodes. Additionally, `machine-controller-manager` includes logic to select machines based on `maxUnavailable/maxSurge`, ensuring machines are rolled in a controlled manner. If only `gardener-node-agent` was used, we would need to implement new logic to track node update progress and to manage the number of nodes updated at a time before proceeding to the next set.

> - `machine-controller-manager`'s drain logic, which offers special handling, cannot be utilized.

Reviewer 2: What exactly is the logic for pods with volumes in MCM? I forgot the details, but I remember we drain pods without volumes in parallel, but pods with volumes sequentially. What exactly is the trigger/condition to continue with the next pod with a volume? In the case of in-place updates, without `maxSurge`, such pods may become pending forever, which is expected and accepted for bare metal machines, so I am asking whether there is some special drain logic we must possibly adapt in MCM in case it would otherwise block (or slow down, because I believe to remember we have some 45s counter or so) the in-place update process.

Author: > we drain pods without volumes in parallel, but pods with volumes sequentially.

Yes, pods with volumes are deleted sequentially and these are steps which is performed while evicting the pods - 
- Eviction or Deletion
- Volume Detachment
- Volume Reattachment

If any of the step fails the pod is added to the retry list and the [logic](https://github.com/gardener/machine-controller-manager/blob/31834e2a9beee5bdc3663cf2cce0d6f63d0c20a4/pkg/util/provider/drain/drain.go#L692-L821) continue with the next pod.

> so I am asking whether there is some special drain logic we must possibly adapt in MCM in case it would otherwise block (or slow down, because I believe to remember we have some 45s counter or so) the in-place update process.

There is no blocking, but surely the process is slow because pods are deleted sequentially and there is a timeout for the steps mentioned above while evicting the pods with volume.
We are currently not looking forward to modifying/enhancing the drain logic as part of this GEP.

Reviewer 2: Right, we cannot track the pod (as it's gone), but we track whether the volume was reattached and if not continue only after the grace period. Thanks!

1. On a shoot spec change for a worker pool, `machine-controller-manager` will label all nodes of the updated worker pool with the `node.machine.sapcloud.io/candidate-for-update` label. This label is used to identify nodes that require an update.
2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent them from being scaled down by [`cluster-autoscaler`](https://github.com/gardener/autoscaler/tree/machine-controller-manager-provider/cluster-autoscaler) during the update process.
3. `machine-controller-manager` will select as many nodes as are permitted by the `maxUnavailable` setting for the worker pool and label it/them with `node.machine.sapcloud.io/selected-for-update`. This label can be used to identify nodes that are about to or are currently undergoing an update.
4. `machine-controller-manager` will cordon and drain the node/nodes and label it/them with the `node.machine.sapcloud.io/ready-for-update` label once the drain is completed.
5. `gardener-node-agent` will detect the `node.machine.sapcloud.io/ready-for-update` label and perform the update on the machine. It will wait for the update to complete (with a specified timeout).

> 6. Once the machine is updated, `gardener-node-agent` will delete all the pods in the node so that they will get recreated and the node will be labelled by the `gardener-node-agent` with the `node.machine.sapcloud.io/update-successful` label.

Reviewer 7: I am a bit confused by this. Do you mind adding details why this is done?

My assumption is that this is so that all pods are for example using the new kubernetes version. However, this should be somewhat aligned, i.e. if we delete the pods, set the label and uncordon the node (next step) in short succession, workload pods might be scheduled before the system pods are ready. This could lead to corresponding failures. For new nodes, we have addressed this with the critical components handling. Do we need something similar here?

Reviewer 3: > I am a bit confused by this. Do you mind adding details why this is done?

This is to make sure that the daemonset pods are recreated.

> Do we need something similar here?

Nice question. We can maybe add the same taints to the node during the update and make GRM remove it later on.

Reviewer 4: This is also a bit unclear to me. Once the node has been drained of all the pods then node agent starts the update. Then what is meant by `gardener-node-agent will delete all the pods in the node so that they will get recreated `?

Reviewer 3: This is after the update is completed by the gardener-node-agent and before MCM uncordones the node. To make sure the daemonset pods are recreated after the update.

2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent them from being scaled down by `cluster-autoscaler` during the update process.
3. `machine-controller-manager` will wait for the `node.machine.sapcloud.io/selected-for-update` label on the node. The user is solely responsible for orchestrating the update and is free to select the nodes to be updated at will.
4. Once the user adds the `node.machine.sapcloud.io/selected-for-update` label, steps 4-7 of the `AutoInPlaceUpdate` strategy will be performed.

![ManualInPlaceUpdate](assets/31-manual-inplace-update.png "Manual In Place Update Strategy")

A new optional [constraint](https://github.com/gardener/gardener/blob/master/docs/usage/shoot/shoot_status.md#constraints) will be introduced in the Shoot status to indicate that some worker pools are undergoing a manual in-place update. Although the Shoot will be marked as successfully reconciled, these worker pools might remain outdated. Furthermore, a field (can be called `pendingWorkersInPlaceUpdate`) will be introduced in the Shoot status to track worker pools awaiting in-place updates.

> Subsequent updates to the worker pool will be blocked by validation if an in-place update is already in progress. This ensures that worker pools do not skip intermediate Kubernetes minor versions or machine image versions. However, if an in-place update fails and a fix - such as a patch to the current updated minor version - is required, it will be allowed, if the Shoot has the `shoot.gardener.cloud/force-in-place-update: true` annotation.

Reviewer 7: What about other changes to the worker pool configuration? Are those possible during an in-place update? For example, if I want to disable running the system components in a worker group or I need to adapt the size of the worker group while a long-running in-place update is ongoing, is this possible or not?
My understanding was that replacement of parts of physical machines can take some arbitrary amount of time. Freezing a worker group for the time if something fails during an in-place update seems like a problem.
WDYT?

Author: > if I want to disable running the system components in a worker group or I need to adapt the size of the worker group while a long-running in-place update is ongoing, is this possible or not?

Other changes which don't require rolling updates, will be allowed.

### Gardener

#### `CloudProfile` API

> A new field `minVersionForInPlaceUpdate` is introduced under `.spec.machineImages[].versions[]` in the `CloudProfile`. This field specifies the minimum machine image version eligible for direct in-place upgrades to a newer version. If the current version is lower than `minVersionForInPlaceUpdate`, the update will be disallowed by validation.

Reviewer 7: Nit: The name `minVersionForInPlaceUpdate` is fairly cumbersome. Would it be possible to shorten it to `minVersionForUpdate`? In case it is not defined, updates are always allowed. If it is specified, it works as described here.
This way it could be used for ordinary image versions as well even if we will likely not use it that way and it is shorter. However, this would be different in case it is not specified.
WDYT?

Author: We can not use it for ordinary images. This field will also be used to determine whether an image supports in-place updates and If an image doesn't specify `minVersionForInPlaceUpdate` field we consider that the image doesn't support in-place updates. So we would like to keep it to `minVersionForInPlaceUpdate`.

What use case do you have in mind for the `minVersionForUpdate` field for ordinary images?

Reviewer 7: I had no particular use case in mind and was primarily concerned about the rather lengthy name. It can stay as is, though.

Reviewer 4: >We can not use it for ordinary images. This field will also be used to determine whether an image supports in-place updates and If an image doesn't specify minVersionForInPlaceUpdate field we consider that the image doesn't support in-place updates. So we would like to keep it to minVersionForInPlaceUpdate.

Piggy-backing on `minVersionForInPlaceUpdate` to determine if an image supports in-place update is not very clean. I agree with @ScheererJ that  minimum version check is a generic check and IMHO for a cleaner API one should have a dedicated field indicating `In-Place OS update support` instead of overloading this field.

Reviewer 3: Do you want to go for something like:
```
inPlaceUpdateConfig:
  supported: true/false
  minVersionForUpdate: <version>
```

```yaml
apiVersion: core.gardener.cloud/v1beta1
kind: CloudProfile
metadata:
  name: cloudprofile1
spec:
  type: <some-provider-name>
  providerConfig: <some-provider-specific-cloudprofile-config>
  kubernetes:
    versions:
      - version: 1.31.1
  machineImages:
    - name: gardenlinux
      versions:
        - version: 1443.8.0
          expirationDate: "2025-02-28T23:59:59Z"
          minVersionForInPlaceUpdate: 1312.3.0
        - version: 1443.7.0 # if minVersionForInPlaceUpdate is not specified, in-place update cannot be performed on this version
  machineTypes:
    - name: m5.large
      cpu: "2"
      gpu: "0"
      memory: 8Gi
      usable: true
  regions:
    - name: <region-name>
```

#### `Shoot` API

> A new field `updateStrategy` is introduced under `spec.provider.workers[]` in the Shoot spec. This field will be passed on to the worker extension. Once set, changing the `updateStrategy` between in-place and replace strategies is prohibited through validation. Additionally, when `AutoInPlaceUpdate` or `ManualInPlaceUpdate` is configured as an update strategy, skipping an intermediate Kubernetes minor version is no longer allowed, because of potential breaking changes that are then not carried out anymore when skipping versions (only allowed with `AutoReplace` as the machines/nodes are created from scratch in this case).

Reviewer 7: Would you mind adding explicitly whether switching between the two in-place update modes is allowed or not?
From this paragraph, it sounds like this is not supported, but in the goals it seems like a goal.

Another question would be if the switch is allowed while an in-place update is ongoing.

Reviewer 3: > Once set, changing the `updateStrategy` between in-place and replace strategies is prohibited through validation.

It's not between the in-place update strategies, but between in-place and replace. We will reword the sentence to:
> Once set, changing the `updateStrategy` from an in-place strategy to a replace strategy and vice versa is prohibited through validation.

> Another question would be if the switch is allowed while an in-place update is ongoing.

We don't see any hurdle to allowing that, but we won't allow it, at least for the first iteration.

  secretRef:
    name: cloudprovider
    namespace: shoot--foobar--aws
  infrastructureProviderStatus: <infra-status>
  pools:
    - name: cpu-worker-1
      machineImage:
        name: gardenlinux
        version: 1.2.2
      minimum: 5
      maximum: 5
      maxSurge: 0
      maxUnavailable: 2
      updateStrategy: AutoInPlaceUpdate # AutoReplaceUpdate/AutoInPlaceUpdate/ManualInPlaceUpdate
      kubelet:
        evictionHard:
          memoryAvailable: 100Mi
          imageFSAvailable: 5%
```

#### `OperatingSystemConfig` API

New fields `osVersion`, `kubeletVersion`, and `credentialsRotation` are introduced in the `OperatingSystemConfig` spec and `inPlaceUpdateConfig` in the `OperatingSystemConfig` status.

- `osVersion` specifies the machine image version.
- `kubeletVersion` specifies the version of the kubelet.
- `credentialsRotation` contains two subfields:
  - `certificateAuthorities` contains `lastInitiationTime` which records the timestamp of the most recent Certificate Authority (CA) rotation initiation.
  - `serviceAccountKey` contains `lastInitiationTime` which records the timestamp of the most recent `ServiceAccount` signing key rotation initiation.

> - `inPlaceUpdateConfig` contains a subfield `osUpdateCommand` which defines the command responsible for performing machine image updates. The command could be invoking an inbuilt utility/tool or a custom script, optionally accompanied by additional arguments or flags like the target version or the OCI registry from which the updated machine image should be pulled, offering flexibility tailored to the needs of different OS extensions.

Reviewer 7: Do we really want to have the command AND arguments as a single string? I would prefer to have it as a slice similar to `pod.spec.containers[].command`. It would prevent any whitespace related issues due to splitting the command at whitespace boundaries.

Reviewer 3: Agreed. I can change the fields to `osUpdateCommand` string and `osUpdateCommandArgs` slice.

### Failures and Recovery Strategy

In-place updates may fail because of several factors ranging from network issues to compatibility problems. This section outlines the failure scenarios, error classifications, and the specific roles `gardener-node-agent` and `machine-controller-manager` play in failure handling and recovery to ensure minimal disruption. The error will be reported in the `node.machine.sapcloud.io/update-failure-message` annotation on the node which can be used subsequently by the Shoot care controller.

#### Types of Failures During Update

##### Drain Timeout/Failure

- A node is marked as `node.machine.sapcloud.io/ready-for-update` only after a successful drain.
- We use the existing [drain logic](https://github.com/gardener/machine-controller-manager/tree/v0.54.0/pkg/util/provider/drain) provided by the `machine-controller-manager`.
- If a drain operation times out or fails, the node will be forcefully drained.

##### Kubernetes Update Failures

> - **Update Halt on Failure:** `gardener-node-agent` will label the failed node with an `node.machine.sapcloud.io/update-failed` label. This label signals `machine-controller-manager` to stop further updates when the number of unhealthy machines reaches configured `maxUnavailable`.

Reviewer 7: What does `Kubernetes Update Failures` mean or in other words what does `gardener-node-agent` check to signal this failure?
I can image "easy" failures, e.g. `kubelet` failing to start, and "harder-to-detect" failures, e.g. pods not being started correctly and hence failing.
Could you please share what you intend to do here?

Author: After the update if GNA shows the kubelet status as unhealthy for a certain period then we will consider that the kubelet update failed.

> "harder-to-detect" failures, e.g. pods not being started correctly and hence failing.

For this, I think it is not possible.  A similar problem is present with `AutoReplaceUpdate` which we use as of today. If on the new node pods don't come up properly we can't do much.

- **Timeout Handling**: After the `machine-controller-manager` adds the `node.machine.sapcloud.io/ready-for-update` label to a node, it waits for the `gardener-node-agent` to complete the update within the timeout configured in `machine-controller-manager`. The timeout on the `machine-controller-manager` side is essential for handling cases where `gardener-node-agent` fails to report any status, such as during a total failure of the OS update. If the update is unsuccessful after a certain number of retries, `machine-controller-manager` will add the `node.machine.sapcloud.io/update-failed` label to the node.
- **Failed Update Handling**: If the nodes labeled with `node.machine.sapcloud.io/update-failed` reaches configured `maxUnavailable`, `machine-controller-manager` stops further updates. The failed machines remain with the old `MachineSet`, and no additional machines in that `MachineDeployment` will be updated until the nodes with the `node.machine.sapcloud.io/update-failed` label are fixed. The failure is propagated to the shoot status, notifying the operator or user, who may need to investigate the issue manually.

> To resolve the issue, the error must be fixed manually, and the `node.machine.sapcloud.io/update-failed` label should be removed from the node. This allows the `machine-controller-manager` to proceed with the update process.

Reviewer 7: What is the effect of removing the `update-failed` label on a node? Is the update retried on the node or do operators somehow perform the update manually?

Author: All the in-place update-related labels will already be present even in case of update-failed scenarios.  Once the update-failed label is removed by the user, the normal in-place update flow will be triggered by MCM.

This design was chosen to leverage the `machine-controller-manager`'s node rollout orchestration functionality while ensuring that it remains a standalone component.

### Prerequisites

To perform an in-place OS update, the following prerequisites must be met:

- The node's machine image or operating system must support in-place updates, providing a tool or utility to initiate the update process.
- The update mechanism should ensure reliability by:
  - Booting into the new version upon a successful update.
  - Falling back to the previous version in case of failure.

Additionally, the update tool or utility may offer configuration options, such as:

- Retry Configuration: The ability to define the number of retries for the update process.
- Registry Configuration: The option to specify the registry from which the OS image is pulled.

### Update Strategies

> Gardener will introduce two additional update strategies, `AutoInPlaceUpdate` and `ManualInPlaceUpdate`, for worker pools, in addition to the existing rolling update which will be now named as `AutoReplaceUpdate` strategy. In the future, there could also be `ManualReplaceUpdate` which could give more control to the users in rolling updates as well. The two new strategies are described as follows:

Reviewer 4: Any reason why the existing rolling update should be renamed to `AutoReplaceUpdate`. This term is not used anywhere in k8s to represent a rolling update and its also quite confusing. Can you explain the rational to make this change?

Author: Rolling refers to replacing instances of a previous version with a newer version in a controlled manner. While **AutoReplaceUpdate** could technically be named **RollingUpdate**, the term "rolling update" simply implies an update process performed in a controlled manner—it doesn’t specify whether nodes are deleted or not.  

On the other hand, **AutoInPlaceUpdate** also updates dependencies on a node in a controlled manner, so it could have been named **InPlaceRollingUpdate**. 

However, for people familiar with Kubernetes, "rolling update" often implies updating to a newer version while deleting the old version. In the case of **InPlaceRollingUpdate**, nothing is being deleted, which could lead to confusion.  

So ideally we needed to avoid the words `rolling update` cause otherwise it would be required to add in both `AutoReplaceUpdate` and `AutoInPlaceUpdate` and in one case old nodes are being deleted and in other old nodes are not being deleted. This can lead to confusion.

So to make the strategies explicitly clear through their names, **AutoReplaceUpdate** seems like a better choice as it directly conveys that the old node will be replaced.

Reviewer 4: Rolling update in MCM is used along with `MaxSurge` and `MaxUnavailable` configuration. Therefore it is not strictly `Replace`. This is the reason why its just called `RollingUpdate` because `MaxSurge` and `MaxUnavailable` determines if will replace or can also create additional VMs.

2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent them from being scaled down by [`cluster-autoscaler`](https://github.com/gardener/autoscaler/tree/machine-controller-manager-provider/cluster-autoscaler) during the update process.
3. `machine-controller-manager` will select as many nodes as are permitted by the `maxUnavailable` setting for the worker pool and label it/them with `node.machine.sapcloud.io/selected-for-update`. This label can be used to identify nodes that are about to or are currently undergoing an update.
4. `machine-controller-manager` will cordon and drain the node/nodes and label it/them with the `node.machine.sapcloud.io/ready-for-update` label once the drain is completed.
5. `gardener-node-agent` will detect the `node.machine.sapcloud.io/ready-for-update` label and perform the update on the machine. It will wait for the update to complete (with a specified timeout).
6. Once the machine is updated, `gardener-node-agent` will delete all the pods in the node so that they will get recreated and the node will be labelled by the `gardener-node-agent` with the `node.machine.sapcloud.io/update-successful` label.

> 7. `machine-controller-manager` will `uncordon` any node with the `node.machine.sapcloud.io/update-successful` label and remove this and all other update-related labels. Once the node becomes `Ready`, it can host workload again.

Reviewer 4: If the update to a newer version is not successful and the update tool has reverted the OS version to the older version (that existed prior to the update) then its still effectively functional. Should it remain `cordoned` and unavailable as a target for the kube-scheduler?

Reviewer 3: Yes, let's make it so. otherwise, we will have to go through the drain again. Anyways, if the failed nodes reach `maxUnvailable`, MCM stops the update. This is also what happens in the case of normal updates as well, right?

Reviewer 4: ok

  kubernetes:
    version: 1.27.3
  networking:
    type: <some-network-extension-name> # {calico,cilium}
```

#### `Worker` API

New field `updateStrategy` and `kubelet` are introduced under `.spec.pools[]` in the Worker API. The `updateStrategy` field will be used to set the `machineDeploymentStrategy` for the `machinev1alpha1.MachineDeployment` resource and `kubelet` field corresponds to the kubelet configuration for this worker pool.

```yaml
apiVersion: extensions.gardener.cloud/v1alpha1
kind: Worker
metadata:
  name: worker
  namespace: shoot--foobar--aws
spec:
  type: aws
  region: eu-west-1
  secretRef:
    name: cloudprovider
    namespace: shoot--foobar--aws
  infrastructureProviderStatus: <infra-status>
  pools:
    - name: cpu-worker-1
      machineImage:
        name: gardenlinux
        version: 1.2.2
      minimum: 5
      maximum: 5
      maxSurge: 0
      maxUnavailable: 2
      updateStrategy: AutoInPlaceUpdate # AutoReplaceUpdate/AutoInPlaceUpdate/ManualInPlaceUpdate
      kubelet:
        evictionHard:
          memoryAvailable: 100Mi
          imageFSAvailable: 5%
```

#### `OperatingSystemConfig` API

> New fields `osVersion`, `kubeletVersion`, and `credentialsRotation` are introduced in the `OperatingSystemConfig` spec and `inPlaceUpdateConfig` in the `OperatingSystemConfig` status.

Reviewer 4: OperatingSystemConfig is no longer just OS config, right? Are there are any plans to either split or rename it?

Reviewer 3: > Are there are any plans to either split or rename it?

There aren't any plans as of now.

2. **Shallow Delete**: Instead of directly moving a `Machine` from one `MachineSet` to another, first perform a shallow delete of the old `Machine` object (without deleting the underlying physical/virtual machine), then create a new `Machine` object in the new `MachineSet` (for the same physical/virtual machine).

> [!NOTE]
> At the time of writing this GEP, it is undecided which method to proceed with. The first method has been evaluated, but the feasibility of the second method is yet to be assessed. If feasible, the second method is preferred.

#### `MachineDeployment` API

A new strategy type `InPlaceUpdate` and a new field, `inPlaceUpdate`, are introduced under the `.spec.strategy` field. `inPlaceUpdate` can be configured when the selected strategy type is `InPlaceUpdate`.

```yaml
apiVersion: machine.sapcloud.io/v1alpha1
kind: MachineDeployment
metadata:
  name: <name>
  namespace: <shoot-namespace>
spec:
  replicas: 1
  selector:
    matchLabels:
      name: <label>
  strategy:
    inPlaceUpdate:
      maxUnavailable: 1
      maxSurge: 0

>       manualUpdate: false # true/false, defaulted to false

Reviewer 5: Thanks for all the changes. A minor question on this field `manualUpdate`. Do you think there may be further additions to the orchestration behaviour in the future - anything that further refines the basic manual or automatic options ?

If so, for future compatibility it might be better to make it this an enum like `orchestrationType:  <enumVariant>` instead of a basic `bool`. If no such varying behaviour is envisaged, then the `bool` is OK.

![AutoInPlaceUpdate](assets/31-auto-inplace-update.png "Auto In Place Update Strategy")

#### `ManualInPlaceUpdate` strategy

1. On a shoot spec change for a worker pool, `machine-controller-manager` will label all nodes of the changed worker pool with the `node.machine.sapcloud.io/candidate-for-update` label. This label is used to identify nodes that require an update.
2. `machine-controller-manager` will add the necessary annotation to the nodes to prevent them from being scaled down by `cluster-autoscaler` during the update process.
3. `machine-controller-manager` will wait for the `node.machine.sapcloud.io/selected-for-update` label on the node. The orchestration of the update is solely the responsibility of the user. The user is free to select the nodes to be updated at will.
4. Once the `node.machine.sapcloud.io/selected-for-update` label is added by the user, steps 4-7 of the `AutoInPlaceUpdate` strategy will be performed.

![ManualInPlaceUpdate](assets/31-manual-inplace-update.png "Manual In Place Update Strategy")

> A new optional [constraint](https://github.com/gardener/gardener/blob/master/docs/usage/shoot/shoot_status.md#constraints) will be introduced in the Shoot status to indicate that some worker pools are undergoing a manual in-place update. While the Shoot will be marked as successfully reconciled, these worker pools may still be outdated. A new field will be added to the Shoot status to keep track of the pending worker pools.

Reviewer 8: > A new field will be added to the Shoot status to keep track of the pending worker pools.

Can you make this more specific and show an example how this looks like? Perhaps you also want to link to #11027 which introduces the API you plan to augment?

  name: pool-01-original
  namespace: default
spec:
  type: gardenlinux
  osVersion: 1631.0
  kubeletVersion: 1.30.4
  credentialRotation:
    caRotationLastInitiationTime:

>     serviceAccountKeyRotationLastInitiationTime:

Reviewer 8: Let's rather follow the structure that already exists in the shoot status

  units:
    - ...
  files:
    - ...
status:
  inPlaceUpdateConfig:
    osUpdateScriptPath: /opt/gardener/bin/inplace-update.sh
    osUpdateScriptArgs:
      - --version
      - 1631.0
      - --repo

>       - <someOCIregistry>

Reviewer 8: What is the benefit of having this in two separate fields? GNA will just always "blindly" call it, or?

Reviewer 3: > GNA will just always "blindly" call it, or?

Yeah, OS extensions are free to use any flags of their choice.

Reviewer 8: OK, so what is the point of having two fields for this (instead of one where the OS extension just puts the entire command with all flags)?

Reviewer 3: Fair enough, updated the doc.

For Kubelet minor version or configuration updates, `gardener-node-agent` will apply the changed files and restart the `kubelet` unit and there are no additional steps involved.

For credential rotation, in the case of Certificate Authority (CA) rotation, the `kubelet` needs to be bootstrapped again so that it can request client certificates signed by the new CA. `gardener-node-agent` will use the existing kubeconfig used by the `kubelet` (passed in the `--kubeconfig` flag. See [`kubelet` Options](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/)) and use it as bootstrap kubeconfig (`--bootstrap-kubeconfig` flag in the kubelet). It will remove the kubelet certificate directory (`--cert-dir` flag in the kubelet) `/var/lib/kubelet/pki`, and the `kubelet.service` unit is restarted. This makes the kubelet request new client certificates through [Certificate signing requests (CSR)](https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#certificate-signing-requests).

> In case of `ServiceAccount` signing key rotation, `gardener-node-agent` will restart itself to load the new service account token using its [`token` controller](https://github.com/gardener/gardener/blob/master/docs/concepts/node-agent.md#token-controller).

Reviewer 8: Is a restart really needed?

Reviewer 3: We can also add a watch for the secret with the token and update it instantly.

Reviewer 8: Once upon a time, we had a watch. However, we removed it as part of https://github.com/gardener/gardener/pull/8857. However, since the node-agent knows when to re-read it (after performing the update), we could make it trigger the controller (via a channel source or so). This way, we can avoid the restart.

Reviewer 3: Sure. Makes sense.
